

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Analysis &mdash; ClinicalKnowledgeGraph 1.0.0 documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="../_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script type="text/javascript" src="../_static/jquery.js"></script>
        <script type="text/javascript" src="../_static/underscore.js"></script>
        <script type="text/javascript" src="../_static/doctools.js"></script>
        <script type="text/javascript" src="../_static/language_data.js"></script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Vizualization" href="analytics_core.viz.html" />
    <link rel="prev" title="Analytics Core" href="analytics_core.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html" class="icon icon-home"> ClinicalKnowledgeGraph
          

          
          </a>

          
            
            
              <div class="version">
                1.0.0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../INTRO.html">Clinical Knowledge Graph</a></li>
</ul>
<p class="caption"><span class="caption-text">First steps</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../intro/getting-started-with-neo4j.html">Getting Started with Neo4j</a></li>
<li class="toctree-l1"><a class="reference internal" href="../intro/getting-started-with-build.html">Getting Started with the CKG Build</a></li>
</ul>
<p class="caption"><span class="caption-text">Getting started</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../getting_started/connect-to-ckg.html">Connecting to the Clinical Knowledge Graph database</a></li>
<li class="toctree-l1"><a class="reference internal" href="../getting_started/create-new-user.html">Create a new user in the graph database</a></li>
<li class="toctree-l1"><a class="reference internal" href="../getting_started/create-new-project.html">Create a new project in the database</a></li>
<li class="toctree-l1"><a class="reference internal" href="../getting_started/upload-data.html">Upload project experimental data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../getting_started/data-analysis-config.html">Define data analysis parameters</a></li>
<li class="toctree-l1"><a class="reference internal" href="../getting_started/access-report.html">Accessing the analysis report</a></li>
<li class="toctree-l1"><a class="reference internal" href="../getting_started/notifications.html">Report notifications</a></li>
</ul>
<p class="caption"><span class="caption-text">Project Report</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../project_report/project-report.html">Generating a project report</a></li>
<li class="toctree-l1"><a class="reference internal" href="../project_report/project-tabs.html">Project report tabs</a></li>
</ul>
<p class="caption"><span class="caption-text">CKG Builder</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../ckg_builder/ontologies.html">Ontology sources and raw file parsers</a></li>
<li class="toctree-l1"><a class="reference internal" href="../ckg_builder/databases.html">Biomedical databases and resources</a></li>
<li class="toctree-l1"><a class="reference internal" href="../ckg_builder/experiments.html">Parsing experimental data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../ckg_builder/graphdb-builder.html">Building the graph database from one Python module</a></li>
</ul>
<p class="caption"><span class="caption-text">Advanced Features</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../advanced_features/import-statistics.html">Clinical Knowledge Graph Statistics: Imports</a></li>
<li class="toctree-l1"><a class="reference internal" href="../advanced_features/graphdb-statistics.html">Clinical Knowledge Graph Statistics: Database</a></li>
<li class="toctree-l1"><a class="reference internal" href="../advanced_features/ckg-notebooks.html">Using Jupyter Notebooks with the Clinical Knowledge Graph</a></li>
<li class="toctree-l1"><a class="reference internal" href="../advanced_features/ckg-queries.html">Retrieving data from the Clinical Knowledge Graph database</a></li>
<li class="toctree-l1"><a class="reference internal" href="../advanced_features/standard-analysis.html">Standardizing the data analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../advanced_features/visualization-plots.html">Visualization plots</a></li>
<li class="toctree-l1"><a class="reference internal" href="../advanced_features/R-interface.html">Python - R interface</a></li>
</ul>
<p class="caption"><span class="caption-text">System Requirements</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../system_require/mac-os.html">Mac OS X requirements</a></li>
<li class="toctree-l1"><a class="reference internal" href="../system_require/linux.html">Linux requirements</a></li>
<li class="toctree-l1"><a class="reference internal" href="../system_require/windows.html">Windows requirements</a></li>
</ul>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="../src.html">Clinical Knowledge Graph package</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="graphdb_connector.html">Graph Database Connector</a></li>
<li class="toctree-l2"><a class="reference internal" href="graphdb_builder.html">Graph Database Builder</a></li>
<li class="toctree-l2"><a class="reference internal" href="report_manager.html">Report Manager</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="analytics_core.html">Analytics Core</a><ul class="current">
<li class="toctree-l3 current"><a class="current reference internal" href="#">Analysis</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#module-analytics_core.analytics.analytics">Analytics</a></li>
<li class="toctree-l4"><a class="reference internal" href="#module-analytics_core.analytics.wgcnaAnalysis">WGCNA analytics</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="analytics_core.viz.html">Vizualization</a></li>
<li class="toctree-l3"><a class="reference internal" href="analytics_core.html#module-analytics_core.R_wrapper">R wrapper</a></li>
<li class="toctree-l3"><a class="reference internal" href="analytics_core.html#module-analytics_core.analytics_factory">Analytics factory</a></li>
<li class="toctree-l3"><a class="reference internal" href="analytics_core.html#module-analytics_core.utils">Utils</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="notebooks.html">Notebooks</a></li>
</ul>
</li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../MANIFEST.html">Credits</a></li>
<li class="toctree-l1"><a class="reference internal" href="../MANIFEST.html#backers">Backers</a></li>
<li class="toctree-l1"><a class="reference internal" href="../MANIFEST.html#contributing">Contributing</a></li>
<li class="toctree-l1"><a class="reference internal" href="../MANIFEST.html#history">History</a></li>
<li class="toctree-l1"><a class="reference internal" href="../MANIFEST.html#code-of-conduct">Code of Conduct</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">ClinicalKnowledgeGraph</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html">Docs</a> &raquo;</li>
        
          <li><a href="../src.html">Clinical Knowledge Graph package</a> &raquo;</li>
        
          <li><a href="analytics_core.html">Analytics Core</a> &raquo;</li>
        
      <li>Analysis</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../_sources/_autosummary/analytics_core.analytics.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="analysis">
<h1>Analysis<a class="headerlink" href="#analysis" title="Permalink to this headline">¶</a></h1>
<div class="section" id="module-analytics_core.analytics.analytics">
<span id="analytics"></span><h2>Analytics<a class="headerlink" href="#module-analytics_core.analytics.analytics" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="analytics_core.analytics.analytics.transform_into_wide_format">
<code class="sig-prename descclassname">analytics_core.analytics.analytics.</code><code class="sig-name descname">transform_into_wide_format</code><span class="sig-paren">(</span><em class="sig-param">data</em>, <em class="sig-param">index</em>, <em class="sig-param">columns</em>, <em class="sig-param">values</em>, <em class="sig-param">extra=[]</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/analytics_core/analytics/analytics.html#transform_into_wide_format"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#analytics_core.analytics.analytics.transform_into_wide_format" title="Permalink to this definition">¶</a></dt>
<dd><p>This function converts a Pandas DataFrame from long to wide format using pandas pivot_table() function.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> – long-format Pandas DataFrame</p></li>
<li><p><strong>index</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#list" title="(in Python v3.5)"><em>list</em></a>) – columns that will be converted into the index</p></li>
<li><p><strong>columns</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – column name whose unique values will become the new column names</p></li>
<li><p><strong>values</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – column to aggregate</p></li>
<li><p><strong>extra</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#list" title="(in Python v3.5)"><em>list</em></a>) – additional columns to be kept as columns</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Wide-format pandas DataFrame</p>
</dd>
</dl>
<p>Example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">result</span> <span class="o">=</span> <span class="n">transform_into_wide_format</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="s1">&#39;index&#39;</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="s1">&#39;x&#39;</span><span class="p">,</span> <span class="n">values</span><span class="o">=</span><span class="s1">&#39;y&#39;</span><span class="p">,</span> <span class="n">extra</span><span class="o">=</span><span class="s1">&#39;group&#39;</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="analytics_core.analytics.analytics.transform_into_long_format">
<code class="sig-prename descclassname">analytics_core.analytics.analytics.</code><code class="sig-name descname">transform_into_long_format</code><span class="sig-paren">(</span><em class="sig-param">data, drop_columns, group, columns=['name', 'y']</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/analytics_core/analytics/analytics.html#transform_into_long_format"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#analytics_core.analytics.analytics.transform_into_long_format" title="Permalink to this definition">¶</a></dt>
<dd><p>Converts a Pandas DataDrame from wide to long format using pd.melt() function.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> – wide-format Pandas DataFrame</p></li>
<li><p><strong>drop_columns</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#list" title="(in Python v3.5)"><em>list</em></a>) – columns to be deleted</p></li>
<li><p><strong>group</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a><em> or </em><a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#list" title="(in Python v3.5)"><em>list</em></a>) – column(s) to use as identifier variables</p></li>
<li><p><strong>columns</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#list" title="(in Python v3.5)"><em>list</em></a>) – names to use for the 1)variable column, and for the 2)value column</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Long-format Pandas DataFrame.</p>
</dd>
</dl>
<p>Example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">result</span> <span class="o">=</span> <span class="n">transform_into_long_format</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">drop_columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;sample&#39;</span><span class="p">,</span> <span class="s1">&#39;subject&#39;</span><span class="p">],</span> <span class="n">group</span><span class="o">=</span><span class="s1">&#39;group&#39;</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;name&#39;</span><span class="p">,</span><span class="s1">&#39;y&#39;</span><span class="p">])</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="analytics_core.analytics.analytics.get_ranking_with_markers">
<code class="sig-prename descclassname">analytics_core.analytics.analytics.</code><code class="sig-name descname">get_ranking_with_markers</code><span class="sig-paren">(</span><em class="sig-param">data</em>, <em class="sig-param">drop_columns</em>, <em class="sig-param">group</em>, <em class="sig-param">columns</em>, <em class="sig-param">list_markers</em>, <em class="sig-param">annotation={}</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/analytics_core/analytics/analytics.html#get_ranking_with_markers"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#analytics_core.analytics.analytics.get_ranking_with_markers" title="Permalink to this definition">¶</a></dt>
<dd><p>This function creates a long-format dataframe with features and values to be plotted together with disease biomarker annotations.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> – wide-format Pandas DataFrame with samples as rows and features as columns</p></li>
<li><p><strong>drop_columns</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#list" title="(in Python v3.5)"><em>list</em></a>) – columns to be deleted</p></li>
<li><p><strong>group</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – column to use as identifier variables</p></li>
<li><p><strong>columns</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#list" title="(in Python v3.5)"><em>list</em></a>) – names to use for the 1)variable column, and for the 2)value column</p></li>
<li><p><strong>list_markers</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#list" title="(in Python v3.5)"><em>list</em></a>) – list of features from data, known to be markers associated to disease.</p></li>
<li><p><strong>annotation</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#dict" title="(in Python v3.5)"><em>dict</em></a>) – markers, from list_markers, and associated diseases.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Long-format pandas DataFrame with group identifiers as rows and columns: ‘name’ (identifier), ‘y’ (LFQ intensity), ‘symbol’ and ‘size’.</p>
</dd>
</dl>
<p>Example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">result</span> <span class="o">=</span> <span class="n">get_ranking_with_markers</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">drop_columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;sample&#39;</span><span class="p">,</span> <span class="s1">&#39;subject&#39;</span><span class="p">],</span> <span class="n">group</span><span class="o">=</span><span class="s1">&#39;group&#39;</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;name&#39;</span><span class="p">,</span> <span class="s1">&#39;y&#39;</span><span class="p">],</span> <span class="n">list_markers</span><span class="p">,</span> <span class="n">annotation</span><span class="o">=</span><span class="p">{})</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="analytics_core.analytics.analytics.extract_number_missing">
<code class="sig-prename descclassname">analytics_core.analytics.analytics.</code><code class="sig-name descname">extract_number_missing</code><span class="sig-paren">(</span><em class="sig-param">data, min_valid, drop_cols=['sample'], group='group'</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/analytics_core/analytics/analytics.html#extract_number_missing"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#analytics_core.analytics.analytics.extract_number_missing" title="Permalink to this definition">¶</a></dt>
<dd><p>Counts how many valid values exist in each column and filters column labels with more valid values than the minimum threshold defined.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> – pandas DataFrame with group as rows and protein identifier as column.</p></li>
<li><p><strong>group</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – column label containing group identifiers. If None, number of valid values is counted across all samples, otherwise is counted per unique group identifier.</p></li>
<li><p><strong>min_valid</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/functions.html#int" title="(in Python v3.5)"><em>int</em></a>) – minimum number of valid values to be filtered.</p></li>
<li><p><strong>drop_columns</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#list" title="(in Python v3.5)"><em>list</em></a>) – column labels to be dropped.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>List of column labels above the threshold.</p>
</dd>
</dl>
<p>Example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">result</span> <span class="o">=</span> <span class="n">extract_number_missing</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">min_valid</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">drop_cols</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;sample&#39;</span><span class="p">],</span> <span class="n">group</span><span class="o">=</span><span class="s1">&#39;group&#39;</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="analytics_core.analytics.analytics.extract_percentage_missing">
<code class="sig-prename descclassname">analytics_core.analytics.analytics.</code><code class="sig-name descname">extract_percentage_missing</code><span class="sig-paren">(</span><em class="sig-param">data, missing_max, drop_cols=['sample'], group='group'</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/analytics_core/analytics/analytics.html#extract_percentage_missing"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#analytics_core.analytics.analytics.extract_percentage_missing" title="Permalink to this definition">¶</a></dt>
<dd><p>Extracts ratio of missing/valid values in each column and filters column labels with lower ratio than the minimum threshold defined.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> – pandas dataframe with group as rows and protein identifier as column.</p></li>
<li><p><strong>group</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – column label containing group identifiers. If None, ratio is calculated across all samples, otherwise is calculated per unique group identifier.</p></li>
<li><p><strong>missing_max</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/functions.html#float" title="(in Python v3.5)"><em>float</em></a>) – maximum ratio of missing/valid values to be filtered.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>List of column labels below the threshold.</p>
</dd>
</dl>
<p>Example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">result</span> <span class="o">=</span> <span class="n">extract_percentage_missing</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">missing_max</span><span class="o">=</span><span class="mf">0.3</span><span class="p">,</span> <span class="n">drop_cols</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;sample&#39;</span><span class="p">],</span> <span class="n">group</span><span class="o">=</span><span class="s1">&#39;group&#39;</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="analytics_core.analytics.analytics.imputation_KNN">
<code class="sig-prename descclassname">analytics_core.analytics.analytics.</code><code class="sig-name descname">imputation_KNN</code><span class="sig-paren">(</span><em class="sig-param">data, drop_cols=['group', 'sample', 'subject'], group='group', cutoff=0.6, alone=True</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/analytics_core/analytics/analytics.html#imputation_KNN"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#analytics_core.analytics.analytics.imputation_KNN" title="Permalink to this definition">¶</a></dt>
<dd><p>k-Nearest Neighbors imputation for pandas dataframes with missing data. For more information visit <a class="reference external" href="https://github.com/iskandr/fancyimpute/blob/master/fancyimpute/knn.py">https://github.com/iskandr/fancyimpute/blob/master/fancyimpute/knn.py</a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> – pandas dataframe with samples as rows and protein identifiers as columns (with additional columns ‘group’, ‘sample’ and ‘subject’).</p></li>
<li><p><strong>group</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – column label containing group identifiers.</p></li>
<li><p><strong>drop_cols</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#list" title="(in Python v3.5)"><em>list</em></a>) – column labels to be dropped. Final dataframe should only have gene/protein/etc identifiers as columns.</p></li>
<li><p><strong>cutoff</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/functions.html#float" title="(in Python v3.5)"><em>float</em></a>) – minimum ratio of missing/valid values required to impute in each column.</p></li>
<li><p><strong>alone</strong> (<em>boolean</em>) – if True removes all columns with any missing values.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Pandas dataframe with samples as rows and protein identifiers as columns.</p>
</dd>
</dl>
<p>Example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">result</span> <span class="o">=</span> <span class="n">imputation_KNN</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">drop_cols</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;group&#39;</span><span class="p">,</span> <span class="s1">&#39;sample&#39;</span><span class="p">,</span> <span class="s1">&#39;subject&#39;</span><span class="p">],</span> <span class="n">group</span><span class="o">=</span><span class="s1">&#39;group&#39;</span><span class="p">,</span> <span class="n">cutoff</span><span class="o">=</span><span class="mf">0.6</span><span class="p">,</span> <span class="n">alone</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="analytics_core.analytics.analytics.imputation_mixed_norm_KNN">
<code class="sig-prename descclassname">analytics_core.analytics.analytics.</code><code class="sig-name descname">imputation_mixed_norm_KNN</code><span class="sig-paren">(</span><em class="sig-param">data, index_cols=['group', 'sample', 'subject'], shift=1.8, nstd=0.3, group='group', cutoff=0.6</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/analytics_core/analytics/analytics.html#imputation_mixed_norm_KNN"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#analytics_core.analytics.analytics.imputation_mixed_norm_KNN" title="Permalink to this definition">¶</a></dt>
<dd><p>Missing values are replaced in two steps: 1) using k-Nearest Neighbors we impute protein columns with a higher ratio of missing/valid values than the defined cutoff,     2) the remaining missing values are replaced by random numbers that are drawn from a normal distribution.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> – pandas dataframe with samples as rows and protein identifiers as columns (with additional columns ‘group’, ‘sample’ and ‘subject’).</p></li>
<li><p><strong>group</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – column label containing group identifiers.</p></li>
<li><p><strong>index_cols</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#list" title="(in Python v3.5)"><em>list</em></a>) – list of column labels to be set as dataframe index.</p></li>
<li><p><strong>shift</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/functions.html#float" title="(in Python v3.5)"><em>float</em></a>) – specifies the amount by which the distribution used for the random numbers is shifted downwards. This is in units of the                         standard deviation of the valid data.</p></li>
<li><p><strong>nstd</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/functions.html#float" title="(in Python v3.5)"><em>float</em></a>) – defines the width of the Gaussian distribution relative to the standard deviation of measured values.                         A value of 0.5 would mean that the width of the distribution used for drawing random numbers is half of the standard deviation of the data.</p></li>
<li><p><strong>cutoff</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/functions.html#float" title="(in Python v3.5)"><em>float</em></a>) – minimum ratio of missing/valid values required to impute in each column.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Pandas dataframe with samples as rows and protein identifiers as columns.</p>
</dd>
</dl>
<p>Example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">result</span> <span class="o">=</span> <span class="n">imputation_mixed_norm_KNN</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">index_cols</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;group&#39;</span><span class="p">,</span> <span class="s1">&#39;sample&#39;</span><span class="p">,</span> <span class="s1">&#39;subject&#39;</span><span class="p">],</span> <span class="n">shift</span> <span class="o">=</span> <span class="mf">1.8</span><span class="p">,</span> <span class="n">nstd</span> <span class="o">=</span> <span class="mf">0.3</span><span class="p">,</span> <span class="n">group</span><span class="o">=</span><span class="s1">&#39;group&#39;</span><span class="p">,</span> <span class="n">cutoff</span><span class="o">=</span><span class="mf">0.6</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="analytics_core.analytics.analytics.imputation_normal_distribution">
<code class="sig-prename descclassname">analytics_core.analytics.analytics.</code><code class="sig-name descname">imputation_normal_distribution</code><span class="sig-paren">(</span><em class="sig-param">data, index_cols=['group', 'sample', 'subject'], shift=1.8, nstd=0.3</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/analytics_core/analytics/analytics.html#imputation_normal_distribution"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#analytics_core.analytics.analytics.imputation_normal_distribution" title="Permalink to this definition">¶</a></dt>
<dd><p>Missing values will be replaced by random numbers that are drawn from a normal distribution. The imputation is done for each sample (across all proteins) separately.
For more information visit <a class="reference external" href="http://www.coxdocs.org/doku.php?id=perseus:user:activities:matrixprocessing:imputation:replacemissingfromgaussian">http://www.coxdocs.org/doku.php?id=perseus:user:activities:matrixprocessing:imputation:replacemissingfromgaussian</a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> – pandas dataframe with samples as rows and protein identifiers as columns (with additional columns ‘group’, ‘sample’ and ‘subject’).</p></li>
<li><p><strong>index_cols</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#list" title="(in Python v3.5)"><em>list</em></a>) – list of column labels to be set as dataframe index.</p></li>
<li><p><strong>shift</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/functions.html#float" title="(in Python v3.5)"><em>float</em></a>) – specifies the amount by which the distribution used for the random numbers is shifted downwards. This is in units of the standard deviation of the valid data.</p></li>
<li><p><strong>nstd</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/functions.html#float" title="(in Python v3.5)"><em>float</em></a>) – defines the width of the Gaussian distribution relative to the standard deviation of measured values. A value of 0.5 would mean that the width of the distribution used for drawing random numbers is half of the standard deviation of the data.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Pandas dataframe with samples as rows and protein identifiers as columns.</p>
</dd>
</dl>
<p>Example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">result</span> <span class="o">=</span> <span class="n">imputation_normal_distribution</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">index_cols</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;group&#39;</span><span class="p">,</span> <span class="s1">&#39;sample&#39;</span><span class="p">,</span> <span class="s1">&#39;subject&#39;</span><span class="p">],</span> <span class="n">shift</span> <span class="o">=</span> <span class="mf">1.8</span><span class="p">,</span> <span class="n">nstd</span> <span class="o">=</span> <span class="mf">0.3</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="analytics_core.analytics.analytics.polish_median_normalization">
<code class="sig-prename descclassname">analytics_core.analytics.analytics.</code><code class="sig-name descname">polish_median_normalization</code><span class="sig-paren">(</span><em class="sig-param">data</em>, <em class="sig-param">max_iter=10</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/analytics_core/analytics/analytics.html#polish_median_normalization"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#analytics_core.analytics.analytics.polish_median_normalization" title="Permalink to this definition">¶</a></dt>
<dd><p>This function iteratively normalizes each sample and each feature to its median until medians converge.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> – </p></li>
<li><p><strong>max_iter</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/functions.html#int" title="(in Python v3.5)"><em>int</em></a>) – number of maximum iterations to prevent infinite loop.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Pandas dataframe.</p>
</dd>
</dl>
<p>Example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">result</span> <span class="o">=</span> <span class="n">polish_median_normalization</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">max_iter</span> <span class="o">=</span> <span class="mi">10</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="analytics_core.analytics.analytics.quantile_normalization">
<code class="sig-prename descclassname">analytics_core.analytics.analytics.</code><code class="sig-name descname">quantile_normalization</code><span class="sig-paren">(</span><em class="sig-param">data</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/analytics_core/analytics/analytics.html#quantile_normalization"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#analytics_core.analytics.analytics.quantile_normalization" title="Permalink to this definition">¶</a></dt>
<dd><p>Applies quantile normalization to each column in pandas dataframe.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>data</strong> – pandas dataframe with features as rows and samples as columns.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Pandas dataframe</p>
</dd>
</dl>
<p>Example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">result</span> <span class="o">=</span> <span class="n">quantile_normalization</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="analytics_core.analytics.analytics.linear_normalization">
<code class="sig-prename descclassname">analytics_core.analytics.analytics.</code><code class="sig-name descname">linear_normalization</code><span class="sig-paren">(</span><em class="sig-param">data</em>, <em class="sig-param">method='l1'</em>, <em class="sig-param">axis=0</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/analytics_core/analytics/analytics.html#linear_normalization"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#analytics_core.analytics.analytics.linear_normalization" title="Permalink to this definition">¶</a></dt>
<dd><p>This function scales input data to a unit norm. For more information visit <a class="reference external" href="https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.normalize.html">https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.normalize.html</a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> – pandas dataframe with samples as rows and features as columns.</p></li>
<li><p><strong>method</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – norm to use to normalize each non-zero sample or non-zero feature (depends on axis).</p></li>
<li><p><strong>axis</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/functions.html#int" title="(in Python v3.5)"><em>int</em></a>) – axis used to normalize the data along. If 1, independently normalize each sample, otherwise (if 0) normalize each feature.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Pandas dataframe</p>
</dd>
</dl>
<p>Example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">result</span> <span class="o">=</span> <span class="n">linear_normalization</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">method</span> <span class="o">=</span> <span class="s2">&quot;l1&quot;</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="mi">0</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="analytics_core.analytics.analytics.remove_group">
<code class="sig-prename descclassname">analytics_core.analytics.analytics.</code><code class="sig-name descname">remove_group</code><span class="sig-paren">(</span><em class="sig-param">data</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/analytics_core/analytics/analytics.html#remove_group"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#analytics_core.analytics.analytics.remove_group" title="Permalink to this definition">¶</a></dt>
<dd><p>Removes column with label ‘group’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>data</strong> – pandas dataframe with one column labelled ‘group’</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Pandas dataframe</p>
</dd>
</dl>
<p>Example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">result</span> <span class="o">=</span> <span class="n">remove_group</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="analytics_core.analytics.analytics.calculate_coefficient_variation">
<code class="sig-prename descclassname">analytics_core.analytics.analytics.</code><code class="sig-name descname">calculate_coefficient_variation</code><span class="sig-paren">(</span><em class="sig-param">values</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/analytics_core/analytics/analytics.html#calculate_coefficient_variation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#analytics_core.analytics.analytics.calculate_coefficient_variation" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute the coefficient of variation, the ratio of the biased standard deviation to the mean, in percentage. For more information visit <a class="reference external" href="https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.variation.html">https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.variation.html</a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>values</strong> (<em>ndarray</em>) – numpy array</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The calculated variation along rows.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>ndarray</p>
</dd>
</dl>
<p>Example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">result</span> <span class="o">=</span> <span class="n">calculate_coefficient_variation</span><span class="p">()</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="analytics_core.analytics.analytics.get_coefficient_variation">
<code class="sig-prename descclassname">analytics_core.analytics.analytics.</code><code class="sig-name descname">get_coefficient_variation</code><span class="sig-paren">(</span><em class="sig-param">data</em>, <em class="sig-param">drop_columns</em>, <em class="sig-param">group</em>, <em class="sig-param">columns</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/analytics_core/analytics/analytics.html#get_coefficient_variation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#analytics_core.analytics.analytics.get_coefficient_variation" title="Permalink to this definition">¶</a></dt>
<dd><p>Extracts the coefficients of variation in each group.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> – pandas dataframe with samples as rows and protein identifiers as columns (with additional columns ‘group’, ‘sample’ and ‘subject’).</p></li>
<li><p><strong>drop_columns</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#list" title="(in Python v3.5)"><em>list</em></a>) – column labels to be dropped from the dataframe</p></li>
<li><p><strong>group</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – column label containing group identifiers.</p></li>
<li><p><strong>columns</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#list" title="(in Python v3.5)"><em>list</em></a>) – names to use for the variable column(s), and for the value column(s)</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Pandas dataframe with columns ‘name’ (protein identifier), ‘x’ (coefficient of variation), ‘y’ (mean) and ‘group’.</p>
</dd>
</dl>
<p>Exmaple:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">result</span> <span class="o">=</span> <span class="n">get_coefficient_variation</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">drop_columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;sample&#39;</span><span class="p">,</span> <span class="s1">&#39;subject&#39;</span><span class="p">],</span> <span class="n">group</span><span class="o">=</span><span class="s1">&#39;group&#39;</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="analytics_core.analytics.analytics.get_proteomics_measurements_ready">
<code class="sig-prename descclassname">analytics_core.analytics.analytics.</code><code class="sig-name descname">get_proteomics_measurements_ready</code><span class="sig-paren">(</span><em class="sig-param">df, index_cols=['group', 'sample', 'subject'], drop_cols=['sample'], group='group', identifier='identifier', extra_identifier='name', imputation=True, method='distribution', missing_method='percentage', missing_per_group=True, missing_max=0.3, min_valid=1, value_col='LFQ_intensity'</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/analytics_core/analytics/analytics.html#get_proteomics_measurements_ready"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#analytics_core.analytics.analytics.get_proteomics_measurements_ready" title="Permalink to this definition">¶</a></dt>
<dd><p>Processes proteomics data extracted from the database: 1) filter proteins with high number of missing values (&gt; missing_max or min_valid), 2) impute missing values.
For more information on imputation method visit <a class="reference external" href="http://www.coxdocs.org/doku.php?id=perseus:user:activities:matrixprocessing:filterrows:filtervalidvaluesrows">http://www.coxdocs.org/doku.php?id=perseus:user:activities:matrixprocessing:filterrows:filtervalidvaluesrows</a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> – long-format pandas dataframe with columns ‘group’, ‘sample’, ‘subject’, ‘identifier’ (protein), ‘name’ (gene) and ‘LFQ_intensity’.</p></li>
<li><p><strong>index_cols</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#list" title="(in Python v3.5)"><em>list</em></a>) – column labels to be be kept as index identifiers.</p></li>
<li><p><strong>drop_cols</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#list" title="(in Python v3.5)"><em>list</em></a>) – column labels to be dropped from the dataframe.</p></li>
<li><p><strong>group</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – column label containing group identifiers.</p></li>
<li><p><strong>identifier</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – column label containing feature identifiers.</p></li>
<li><p><strong>extra_identifier</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – column label containing additional protein identifiers (e.g. gene names).</p></li>
<li><p><strong>imputation</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/functions.html#bool" title="(in Python v3.5)"><em>bool</em></a>) – if True performs imputation of missing values.</p></li>
<li><p><strong>method</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – method for missing values imputation (‘KNN’, ‘distribuition’, or ‘mixed’)</p></li>
<li><p><strong>missing_method</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – defines which expression rows are counted to determine if a column has enough valid values to survive the filtering process.</p></li>
<li><p><strong>missing_per_group</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/functions.html#bool" title="(in Python v3.5)"><em>bool</em></a>) – if True filter proteins based on valid values per group; if False filter across all samples.</p></li>
<li><p><strong>missing_max</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/functions.html#float" title="(in Python v3.5)"><em>float</em></a>) – maximum ratio of missing/valid values to be filtered.</p></li>
<li><p><strong>min_valid</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/functions.html#int" title="(in Python v3.5)"><em>int</em></a>) – minimum number of valid values to be filtered.</p></li>
<li><p><strong>value_col</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – column label containing expression values.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Pandas dataframe with samples as rows and protein identifiers (UniprotID~GeneName) as columns (with additional columns ‘group’, ‘sample’ and ‘subject’).</p>
</dd>
</dl>
<p>Example 1:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">result</span> <span class="o">=</span> <span class="n">get_proteomics_measurements_ready</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">index_cols</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;group&#39;</span><span class="p">,</span> <span class="s1">&#39;sample&#39;</span><span class="p">,</span> <span class="s1">&#39;subject&#39;</span><span class="p">],</span> <span class="n">drop_cols</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;sample&#39;</span><span class="p">],</span> <span class="n">group</span><span class="o">=</span><span class="s1">&#39;group&#39;</span><span class="p">,</span> <span class="n">identifier</span><span class="o">=</span><span class="s1">&#39;identifier&#39;</span><span class="p">,</span> <span class="n">extra_identifier</span><span class="o">=</span><span class="s1">&#39;name&#39;</span><span class="p">,</span> <span class="n">imputation</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">method</span> <span class="o">=</span> <span class="s1">&#39;distribution&#39;</span><span class="p">,</span> <span class="n">missing_method</span> <span class="o">=</span> <span class="s1">&#39;percentage&#39;</span><span class="p">,</span> <span class="n">missing_per_group</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">missing_max</span> <span class="o">=</span> <span class="mf">0.3</span><span class="p">,</span> <span class="n">value_col</span><span class="o">=</span><span class="s1">&#39;LFQ_intensity&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>Example 2:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">result</span> <span class="o">=</span> <span class="n">get_proteomics_measurements_ready</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">index_cols</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;group&#39;</span><span class="p">,</span> <span class="s1">&#39;sample&#39;</span><span class="p">,</span> <span class="s1">&#39;subject&#39;</span><span class="p">],</span> <span class="n">drop_cols</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;sample&#39;</span><span class="p">],</span> <span class="n">group</span><span class="o">=</span><span class="s1">&#39;group&#39;</span><span class="p">,</span> <span class="n">identifier</span><span class="o">=</span><span class="s1">&#39;identifier&#39;</span><span class="p">,</span> <span class="n">extra_identifier</span><span class="o">=</span><span class="s1">&#39;name&#39;</span><span class="p">,</span> <span class="n">imputation</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">method</span> <span class="o">=</span> <span class="s1">&#39;mixed&#39;</span><span class="p">,</span> <span class="n">missing_method</span> <span class="o">=</span> <span class="s1">&#39;at_least_x&#39;</span><span class="p">,</span> <span class="n">missing_per_group</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">min_valid</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">value_col</span><span class="o">=</span><span class="s1">&#39;LFQ_intensity&#39;</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="analytics_core.analytics.analytics.get_clinical_measurements_ready">
<code class="sig-prename descclassname">analytics_core.analytics.analytics.</code><code class="sig-name descname">get_clinical_measurements_ready</code><span class="sig-paren">(</span><em class="sig-param">df, subject_id='subject', sample_id='biological_sample', group_id='group', columns=['clinical_variable'], values='values', extra=['group'], imputation=True, imputation_method='KNN'</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/analytics_core/analytics/analytics.html#get_clinical_measurements_ready"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#analytics_core.analytics.analytics.get_clinical_measurements_ready" title="Permalink to this definition">¶</a></dt>
<dd><p>Processes clinical data extracted from the database by converting dataframe to wide-format and imputing missing values.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> – long-format pandas dataframe with columns ‘group’, ‘biological_sample’, ‘subject’, ‘clinical_variable’, ‘value’.</p></li>
<li><p><strong>subject_id</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – column label containing subject identifiers.</p></li>
<li><p><strong>sample_id</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – column label containing biological sample identifiers.</p></li>
<li><p><strong>group_id</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – column label containing group identifiers.</p></li>
<li><p><strong>columns</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#list" title="(in Python v3.5)"><em>list</em></a>) – column name whose unique values will become the new column names</p></li>
<li><p><strong>values</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – column label containing clinical variable values.</p></li>
<li><p><strong>extra</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#list" title="(in Python v3.5)"><em>list</em></a>) – additional column labels to be kept as columns</p></li>
<li><p><strong>imputation</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/functions.html#bool" title="(in Python v3.5)"><em>bool</em></a>) – if True performs imputation of missing values.</p></li>
<li><p><strong>imputation_method</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – method for missing values imputation (‘KNN’, ‘distribuition’, or ‘mixed’).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Pandas dataframe with samples as rows and clinical variables as columns (with additional columns ‘group’, ‘subject’ and ‘biological_sample’).</p>
</dd>
</dl>
<p>Example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">result</span> <span class="o">=</span> <span class="n">get_clinical_measurements_ready</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">subject_id</span><span class="o">=</span><span class="s1">&#39;subject&#39;</span><span class="p">,</span> <span class="n">sample_id</span><span class="o">=</span><span class="s1">&#39;biological_sample&#39;</span><span class="p">,</span> <span class="n">group_id</span><span class="o">=</span><span class="s1">&#39;group&#39;</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;clinical_variable&#39;</span><span class="p">],</span> <span class="n">values</span><span class="o">=</span><span class="s1">&#39;values&#39;</span><span class="p">,</span> <span class="n">extra</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;group&#39;</span><span class="p">],</span> <span class="n">imputation</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">imputation_method</span><span class="o">=</span><span class="s1">&#39;KNN&#39;</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="analytics_core.analytics.analytics.get_summary_data_matrix">
<code class="sig-prename descclassname">analytics_core.analytics.analytics.</code><code class="sig-name descname">get_summary_data_matrix</code><span class="sig-paren">(</span><em class="sig-param">data</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/analytics_core/analytics/analytics.html#get_summary_data_matrix"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#analytics_core.analytics.analytics.get_summary_data_matrix" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns some statistics on the data matrix provided.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>data</strong> – pandas dataframe.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>dictionary with the type of statistics as key and the statistic as value in the shape of a pandas data frame</p>
</dd>
</dl>
<p>Example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">result</span> <span class="o">=</span> <span class="n">get_summary_data_matrix</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="analytics_core.analytics.analytics.check_normality">
<code class="sig-prename descclassname">analytics_core.analytics.analytics.</code><code class="sig-name descname">check_normality</code><span class="sig-paren">(</span><em class="sig-param">data</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/analytics_core/analytics/analytics.html#check_normality"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#analytics_core.analytics.analytics.check_normality" title="Permalink to this definition">¶</a></dt>
<dd><p>Checks whether columns (features) in the provided matrix follow a normal distribution (Requires at least 8 rows). Uses
Pigouin test Normality: <a class="reference external" href="https://pingouin-stats.org/generated/pingouin.normality.html">https://pingouin-stats.org/generated/pingouin.normality.html</a></p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>data</strong> – pandas dataframe.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a pandas data frame with features as rows, test statistic, pvalue, true/false normally distributed. None if</p>
</dd>
</dl>
<p>number of rows below 8.</p>
<p>Example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">result</span> <span class="o">=</span> <span class="n">check_normality</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="analytics_core.analytics.analytics.run_pca">
<code class="sig-prename descclassname">analytics_core.analytics.analytics.</code><code class="sig-name descname">run_pca</code><span class="sig-paren">(</span><em class="sig-param">data, drop_cols=['sample', 'subject'], group='group', components=2, dropna=True</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/analytics_core/analytics/analytics.html#run_pca"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#analytics_core.analytics.analytics.run_pca" title="Permalink to this definition">¶</a></dt>
<dd><p>Performs principal component analysis and returns the values of each component for each sample and each protein, and the loadings for each protein.     For information visit <a class="reference external" href="https://scikit-learn.org/stable/modules/generated/sklearn.decomposition.PCA.html">https://scikit-learn.org/stable/modules/generated/sklearn.decomposition.PCA.html</a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> – pandas dataframe with samples as rows and protein identifiers as columns (with additional columns ‘group’, ‘sample’ and ‘subject’).</p></li>
<li><p><strong>drop_cols</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#list" title="(in Python v3.5)"><em>list</em></a>) – column labels to be dropped from the dataframe.</p></li>
<li><p><strong>group</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – column label containing group identifiers.</p></li>
<li><p><strong>components</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/functions.html#int" title="(in Python v3.5)"><em>int</em></a>) – number of components to keep.</p></li>
<li><p><strong>dropna</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/functions.html#bool" title="(in Python v3.5)"><em>bool</em></a>) – if True removes all columns with any missing values.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Two dictionaries: 1) two pandas dataframes (first one with components values, the second with the components vectors for each protein), 2) xaxis and yaxis titles with components loadings for plotly.</p>
</dd>
</dl>
<p>Example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">result</span> <span class="o">=</span> <span class="n">run_pca</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">drop_cols</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;sample&#39;</span><span class="p">,</span> <span class="s1">&#39;subject&#39;</span><span class="p">],</span> <span class="n">group</span><span class="o">=</span><span class="s1">&#39;group&#39;</span><span class="p">,</span> <span class="n">components</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">dropna</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="analytics_core.analytics.analytics.run_tsne">
<code class="sig-prename descclassname">analytics_core.analytics.analytics.</code><code class="sig-name descname">run_tsne</code><span class="sig-paren">(</span><em class="sig-param">data, drop_cols=['sample', 'subject'], group='group', components=2, perplexity=40, n_iter=1000, init='pca', dropna=True</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/analytics_core/analytics/analytics.html#run_tsne"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#analytics_core.analytics.analytics.run_tsne" title="Permalink to this definition">¶</a></dt>
<dd><p>Performs t-distributed Stochastic Neighbor Embedding analysis. For more information visit <a class="reference external" href="https://scikit-learn.org/stable/modules/generated/sklearn.manifold.TSNE.html">https://scikit-learn.org/stable/modules/generated/sklearn.manifold.TSNE.html</a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> – pandas dataframe with samples as rows and protein identifiers as columns (with additional columns ‘group’, ‘sample’ and ‘subject’).</p></li>
<li><p><strong>drop_cols</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#list" title="(in Python v3.5)"><em>list</em></a>) – column labels to be dropped from the dataframe.</p></li>
<li><p><strong>group</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – column label containing group identifiers.</p></li>
<li><p><strong>components</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/functions.html#int" title="(in Python v3.5)"><em>int</em></a>) – dimension of the embedded space.</p></li>
<li><p><strong>perplexity</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/functions.html#int" title="(in Python v3.5)"><em>int</em></a>) – related to the number of nearest neighbors that is used in other manifold learning algorithms. Consider selecting a value between 5 and 50.</p></li>
<li><p><strong>n_iter</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/functions.html#int" title="(in Python v3.5)"><em>int</em></a>) – maximum number of iterations for the optimization (at least 250).</p></li>
<li><p><strong>init</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – initialization of embedding (‘random’, ‘pca’ or numpy array of shape n_samples x n_components).</p></li>
<li><p><strong>dropna</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/functions.html#bool" title="(in Python v3.5)"><em>bool</em></a>) – if True removes all columns with any missing values.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Two dictionaries: 1) pandas dataframe with embedding vectors, 2) xaxis and yaxis titles for plotly.</p>
</dd>
</dl>
<p>Example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">result</span> <span class="o">=</span> <span class="n">run_tsne</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">drop_cols</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;sample&#39;</span><span class="p">,</span> <span class="s1">&#39;subject&#39;</span><span class="p">],</span> <span class="n">group</span><span class="o">=</span><span class="s1">&#39;group&#39;</span><span class="p">,</span> <span class="n">components</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">perplexity</span><span class="o">=</span><span class="mi">40</span><span class="p">,</span> <span class="n">n_iter</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span> <span class="n">init</span><span class="o">=</span><span class="s1">&#39;pca&#39;</span><span class="p">,</span> <span class="n">dropna</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="analytics_core.analytics.analytics.run_umap">
<code class="sig-prename descclassname">analytics_core.analytics.analytics.</code><code class="sig-name descname">run_umap</code><span class="sig-paren">(</span><em class="sig-param">data, drop_cols=['sample', 'subject'], group='group', n_neighbors=10, min_dist=0.3, metric='cosine', dropna=True</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/analytics_core/analytics/analytics.html#run_umap"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#analytics_core.analytics.analytics.run_umap" title="Permalink to this definition">¶</a></dt>
<dd><p>Performs Uniform Manifold Approximation and Projection. For more information vist <a class="reference external" href="https://umap-learn.readthedocs.io">https://umap-learn.readthedocs.io</a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> – pandas dataframe with samples as rows and protein identifiers as columns (with additional columns ‘group’, ‘sample’ and ‘subject’).</p></li>
<li><p><strong>drop_cols</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#list" title="(in Python v3.5)"><em>list</em></a>) – column labels to be dropped from the dataframe.</p></li>
<li><p><strong>group</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – column label containing group identifiers.</p></li>
<li><p><strong>n_neighbors</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/functions.html#int" title="(in Python v3.5)"><em>int</em></a>) – number of neighboring points used in local approximations of manifold structure.</p></li>
<li><p><strong>min_dist</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/functions.html#float" title="(in Python v3.5)"><em>float</em></a>) – controls how tightly the embedding is allowed compress points together.</p></li>
<li><p><strong>metric</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – metric used to measure distance in the input space.</p></li>
<li><p><strong>dropna</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/functions.html#bool" title="(in Python v3.5)"><em>bool</em></a>) – if True removes all columns with any missing values.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Two dictionaries: 1) pandas dataframe with embedding of the training data in low-dimensional space, 2) xaxis and yaxis titles for plotly.</p>
</dd>
</dl>
<p>Example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">result</span> <span class="o">=</span> <span class="n">run_umap</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">drop_cols</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;sample&#39;</span><span class="p">,</span> <span class="s1">&#39;subject&#39;</span><span class="p">],</span> <span class="n">group</span><span class="o">=</span><span class="s1">&#39;group&#39;</span><span class="p">,</span> <span class="n">n_neighbors</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">min_dist</span><span class="o">=</span><span class="mf">0.3</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="s1">&#39;cosine&#39;</span><span class="p">,</span> <span class="n">dropna</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="analytics_core.analytics.analytics.calculate_correlations">
<code class="sig-prename descclassname">analytics_core.analytics.analytics.</code><code class="sig-name descname">calculate_correlations</code><span class="sig-paren">(</span><em class="sig-param">x</em>, <em class="sig-param">y</em>, <em class="sig-param">method='pearson'</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/analytics_core/analytics/analytics.html#calculate_correlations"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#analytics_core.analytics.analytics.calculate_correlations" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates a Spearman (nonparametric) or a Pearson (parametric) correlation coefficient and p-value to test for non-correlation.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> (<em>ndarray</em>) – array 1</p></li>
<li><p><strong>y</strong> (<em>ndarray</em>) – array 2</p></li>
<li><p><strong>method</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – chooses which kind of correlation method to run</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Tuple with two floats, correlation coefficient and two-tailed p-value.</p>
</dd>
</dl>
<p>Example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">result</span> <span class="o">=</span> <span class="n">calculate_correlations</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;pearson&#39;</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="analytics_core.analytics.analytics.apply_pvalue_fdrcorrection">
<code class="sig-prename descclassname">analytics_core.analytics.analytics.</code><code class="sig-name descname">apply_pvalue_fdrcorrection</code><span class="sig-paren">(</span><em class="sig-param">pvalues</em>, <em class="sig-param">alpha=0.05</em>, <em class="sig-param">method='indep'</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/analytics_core/analytics/analytics.html#apply_pvalue_fdrcorrection"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#analytics_core.analytics.analytics.apply_pvalue_fdrcorrection" title="Permalink to this definition">¶</a></dt>
<dd><p>Performs p-value correction for false discovery rate. For more information visit <a class="reference external" href="https://www.statsmodels.org/devel/generated/statsmodels.stats.multitest.fdrcorrection.html">https://www.statsmodels.org/devel/generated/statsmodels.stats.multitest.fdrcorrection.html</a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>pvalues</strong> (<em>ndarray</em>) – et of p-values of the individual tests.</p></li>
<li><p><strong>alpha</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/functions.html#float" title="(in Python v3.5)"><em>float</em></a>) – error rate.</p></li>
<li><p><strong>method</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – method of p-value correction (‘indep’, ‘negcorr’).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Tuple with two arrays, boolen for rejecting H0 hypothesis and float for adjusted p-value.</p>
</dd>
</dl>
<p>Exmaple:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">result</span> <span class="o">=</span> <span class="n">apply_pvalue_fdrcorrection</span><span class="p">(</span><span class="n">pvalues</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;indep&#39;</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="analytics_core.analytics.analytics.apply_pvalue_twostage_fdrcorrection">
<code class="sig-prename descclassname">analytics_core.analytics.analytics.</code><code class="sig-name descname">apply_pvalue_twostage_fdrcorrection</code><span class="sig-paren">(</span><em class="sig-param">pvalues</em>, <em class="sig-param">alpha=0.05</em>, <em class="sig-param">method='bh'</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/analytics_core/analytics/analytics.html#apply_pvalue_twostage_fdrcorrection"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#analytics_core.analytics.analytics.apply_pvalue_twostage_fdrcorrection" title="Permalink to this definition">¶</a></dt>
<dd><p>Iterated two stage linear step-up procedure with estimation of number of true hypotheses. For more information visit <a class="reference external" href="https://www.statsmodels.org/dev/generated/statsmodels.stats.multitest.fdrcorrection_twostage.html">https://www.statsmodels.org/dev/generated/statsmodels.stats.multitest.fdrcorrection_twostage.html</a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>pvalues</strong> (<em>ndarray</em>) – et of p-values of the individual tests.</p></li>
<li><p><strong>alpha</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/functions.html#float" title="(in Python v3.5)"><em>float</em></a>) – error rate.</p></li>
<li><p><strong>method</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – method of p-value correction (‘bky’, ‘bh’).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Tuple with two arrays, boolen for rejecting H0 hypothesis and float for adjusted p-value.</p>
</dd>
</dl>
<p>Exmaple:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">result</span> <span class="o">=</span> <span class="n">apply_pvalue_twostage_fdrcorrection</span><span class="p">(</span><span class="n">pvalues</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;bh&#39;</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="analytics_core.analytics.analytics.apply_pvalue_permutation_fdrcorrection">
<code class="sig-prename descclassname">analytics_core.analytics.analytics.</code><code class="sig-name descname">apply_pvalue_permutation_fdrcorrection</code><span class="sig-paren">(</span><em class="sig-param">df</em>, <em class="sig-param">observed_pvalues</em>, <em class="sig-param">group</em>, <em class="sig-param">alpha=0.05</em>, <em class="sig-param">permutations=50</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/analytics_core/analytics/analytics.html#apply_pvalue_permutation_fdrcorrection"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#analytics_core.analytics.analytics.apply_pvalue_permutation_fdrcorrection" title="Permalink to this definition">¶</a></dt>
<dd><p>This function applies multiple hypothesis testing correction using a permutation-based false discovery rate approach.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> – pandas dataframe with samples as rows and features as columns.</p></li>
<li><p><strong>oberved_pvalues</strong> – pandas Series with p-values calculated on the originally measured data.</p></li>
<li><p><strong>group</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – name of the column containing group identifiers.</p></li>
<li><p><strong>alpha</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/functions.html#float" title="(in Python v3.5)"><em>float</em></a>) – error rate. Values velow alpha are considered significant.</p></li>
<li><p><strong>permutations</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/functions.html#int" title="(in Python v3.5)"><em>int</em></a>) – number of permutations to be applied.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Pandas dataframe with adjusted p-values and rejected columns.</p>
</dd>
</dl>
<p>Example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">result</span> <span class="o">=</span> <span class="n">apply_pvalue_permutation_fdrcorrection</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">observed_pvalues</span><span class="p">,</span> <span class="n">group</span><span class="o">=</span><span class="s1">&#39;group&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span> <span class="n">permutations</span><span class="o">=</span><span class="mi">50</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="analytics_core.analytics.analytics.get_counts_permutation_fdr">
<code class="sig-prename descclassname">analytics_core.analytics.analytics.</code><code class="sig-name descname">get_counts_permutation_fdr</code><span class="sig-paren">(</span><em class="sig-param">value</em>, <em class="sig-param">random</em>, <em class="sig-param">observed</em>, <em class="sig-param">n</em>, <em class="sig-param">alpha</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/analytics_core/analytics/analytics.html#get_counts_permutation_fdr"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#analytics_core.analytics.analytics.get_counts_permutation_fdr" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates local FDR values (q-values) by computing the fraction of accepted hits from the permuted data over accepted hits from the measured data normalized by the total number of permutations.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>value</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/functions.html#float" title="(in Python v3.5)"><em>float</em></a>) – computed p-value on measured data for a feature.</p></li>
<li><p><strong>random</strong> (<em>ndarray</em>) – p-values computed on the permuted data.</p></li>
<li><p><strong>observed</strong> – pandas Series with p-values calculated on the originally measured data.</p></li>
<li><p><strong>n</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/functions.html#int" title="(in Python v3.5)"><em>int</em></a>) – number of permutations to be applied.</p></li>
<li><p><strong>alpha</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/functions.html#float" title="(in Python v3.5)"><em>float</em></a>) – error rate. Values velow alpha are considered significant.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Tuple with q-value and boolean for H0 rejected.</p>
</dd>
</dl>
<p>Example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">result</span> <span class="o">=</span> <span class="n">get_counts_permutation_fdr</span><span class="p">(</span><span class="n">value</span><span class="p">,</span> <span class="n">random</span><span class="p">,</span> <span class="n">observed</span><span class="p">,</span> <span class="n">n</span><span class="o">=</span><span class="mi">250</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.05</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="analytics_core.analytics.analytics.convertToEdgeList">
<code class="sig-prename descclassname">analytics_core.analytics.analytics.</code><code class="sig-name descname">convertToEdgeList</code><span class="sig-paren">(</span><em class="sig-param">data</em>, <em class="sig-param">cols</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/analytics_core/analytics/analytics.html#convertToEdgeList"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#analytics_core.analytics.analytics.convertToEdgeList" title="Permalink to this definition">¶</a></dt>
<dd><p>This function converts a pandas dataframe to an edge list where index becomes the source nodes and columns the target nodes.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> – pandas dataframe.</p></li>
<li><p><strong>cols</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#list" title="(in Python v3.5)"><em>list</em></a>) – names for dataframe columns.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Pandas dataframe with columns cols.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="analytics_core.analytics.analytics.run_correlation">
<code class="sig-prename descclassname">analytics_core.analytics.analytics.</code><code class="sig-name descname">run_correlation</code><span class="sig-paren">(</span><em class="sig-param">df</em>, <em class="sig-param">alpha=0.05</em>, <em class="sig-param">subject='subject'</em>, <em class="sig-param">group='group'</em>, <em class="sig-param">method='pearson'</em>, <em class="sig-param">correction=('fdr'</em>, <em class="sig-param">'indep')</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/analytics_core/analytics/analytics.html#run_correlation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#analytics_core.analytics.analytics.run_correlation" title="Permalink to this definition">¶</a></dt>
<dd><p>This function calculates pairwise correlations for columns in dataframe, and returns it in the shape of a edge list with ‘weight’ as correlation score, and the ajusted p-values.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> – pandas dataframe with samples as rows and features as columns.</p></li>
<li><p><strong>subject</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – name of column containing subject identifiers.</p></li>
<li><p><strong>group</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – name of column containing group identifiers.</p></li>
<li><p><strong>method</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – method to use for correlation calculation (‘pearson’, ‘spearman’).</p></li>
<li><p><strong>alpha</strong> (<em>floar</em>) – error rate. Values velow alpha are considered significant.</p></li>
<li><p><strong>correction</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#tuple" title="(in Python v3.5)"><em>tuple</em></a>) – first string corresponds to FDR correction type (‘fdr’, ‘2fdr’), and second string determines which method to use (fdr:’indep’, ‘negcorr’, 2fdr:’bky’, ‘bh’).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Pandas dataframe with columns: ‘node1’, ‘node2’, ‘weight’, ‘padj’ and ‘rejected’.</p>
</dd>
</dl>
<p>Example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">result</span> <span class="o">=</span> <span class="n">run_correlation</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span> <span class="n">subject</span><span class="o">=</span><span class="s1">&#39;subject&#39;</span><span class="p">,</span> <span class="n">group</span><span class="o">=</span><span class="s1">&#39;group&#39;</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;pearson&#39;</span><span class="p">,</span> <span class="n">correction</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;fdr&#39;</span><span class="p">,</span> <span class="s1">&#39;indep&#39;</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="analytics_core.analytics.analytics.run_multi_correlation">
<code class="sig-prename descclassname">analytics_core.analytics.analytics.</code><code class="sig-name descname">run_multi_correlation</code><span class="sig-paren">(</span><em class="sig-param">df, alpha=0.05, subject='subject', on=['subject', 'biological_sample'], group='group', method='pearson', correction=('fdr', 'indep')</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/analytics_core/analytics/analytics.html#run_multi_correlation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#analytics_core.analytics.analytics.run_multi_correlation" title="Permalink to this definition">¶</a></dt>
<dd><p>This function merges all input dataframes and calculates pairwise correlations for all columns.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#dict" title="(in Python v3.5)"><em>dict</em></a>) – dictionary of pandas dataframes with samples as rows and features as columns.</p></li>
<li><p><strong>subject</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – name of the column containing subject identifiers.</p></li>
<li><p><strong>group</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – name of the column containing group identifiers.</p></li>
<li><p><strong>on</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#list" title="(in Python v3.5)"><em>list</em></a>) – column names to join dataframes on (must be found in all dataframes).</p></li>
<li><p><strong>method</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – method to use for correlation calculation (‘pearson’, ‘spearman’).</p></li>
<li><p><strong>alpha</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/functions.html#float" title="(in Python v3.5)"><em>float</em></a>) – error rate. Values velow alpha are considered significant.</p></li>
<li><p><strong>correction</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#tuple" title="(in Python v3.5)"><em>tuple</em></a>) – first string corresponds to FDR correction type (‘fdr’, ‘2fdr’), and second string determines which method to use (fdr:’indep’, ‘negcorr’, 2fdr:’bky’, ‘bh’).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Pandas dataframe with columns: ‘node1’, ‘node2’, ‘weight’, ‘padj’ and ‘rejected’.</p>
</dd>
</dl>
<p>Example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">result</span> <span class="o">=</span> <span class="n">run_multi_correlation</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span> <span class="n">subject</span><span class="o">=</span><span class="s1">&#39;subject&#39;</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;subject&#39;</span><span class="p">,</span> <span class="s1">&#39;biological_sample&#39;</span><span class="p">]</span> <span class="p">,</span> <span class="n">group</span><span class="o">=</span><span class="s1">&#39;group&#39;</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;pearson&#39;</span><span class="p">,</span> <span class="n">correction</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;fdr&#39;</span><span class="p">,</span> <span class="s1">&#39;indep&#39;</span><span class="p">))</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="analytics_core.analytics.analytics.calculate_rm_correlation">
<code class="sig-prename descclassname">analytics_core.analytics.analytics.</code><code class="sig-name descname">calculate_rm_correlation</code><span class="sig-paren">(</span><em class="sig-param">df</em>, <em class="sig-param">x</em>, <em class="sig-param">y</em>, <em class="sig-param">subject</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/analytics_core/analytics/analytics.html#calculate_rm_correlation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#analytics_core.analytics.analytics.calculate_rm_correlation" title="Permalink to this definition">¶</a></dt>
<dd><p>Computes correlation and p-values between two columns a and b in df.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> – pandas dataframe with subjects as rows and two features and columns.</p></li>
<li><p><strong>x</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – feature a name.</p></li>
<li><p><strong>y</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – feature b name.</p></li>
<li><p><strong>subject</strong> – column name containing the covariate variable.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Tuple with values for: feature a, feature b, correlation, p-value and degrees of freedom.</p>
</dd>
</dl>
<p>Example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">result</span> <span class="o">=</span> <span class="n">calculate_rm_correlation</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">x</span><span class="o">=</span><span class="s1">&#39;feature a&#39;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s1">&#39;feature b&#39;</span><span class="p">,</span> <span class="n">subject</span><span class="o">=</span><span class="s1">&#39;subject&#39;</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="analytics_core.analytics.analytics.run_rm_correlation">
<code class="sig-prename descclassname">analytics_core.analytics.analytics.</code><code class="sig-name descname">run_rm_correlation</code><span class="sig-paren">(</span><em class="sig-param">df</em>, <em class="sig-param">alpha=0.05</em>, <em class="sig-param">subject='subject'</em>, <em class="sig-param">correction=('fdr'</em>, <em class="sig-param">'indep')</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/analytics_core/analytics/analytics.html#run_rm_correlation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#analytics_core.analytics.analytics.run_rm_correlation" title="Permalink to this definition">¶</a></dt>
<dd><p>Computes pairwise repeated measurements correlations for all columns in dataframe, and returns results as an edge list with ‘weight’ as correlation score, p-values, degrees of freedom and ajusted p-values.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> – pandas dataframe with samples as rows and features as columns.</p></li>
<li><p><strong>subject</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – name of column containing subject identifiers.</p></li>
<li><p><strong>alpha</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/functions.html#float" title="(in Python v3.5)"><em>float</em></a>) – error rate. Values velow alpha are considered significant.</p></li>
<li><p><strong>correction</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#tuple" title="(in Python v3.5)"><em>tuple</em></a>) – first string corresponds to FDR correction type (‘fdr’, ‘2fdr’), and second string determines which method to use (fdr:’indep’, ‘negcorr’, 2fdr:’bky’, ‘bh’).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Pandas dataframe with columns: ‘node1’, ‘node2’, ‘weight’, ‘pvalue’, ‘dof’, ‘padj’ and ‘rejected’.</p>
</dd>
</dl>
<p>Example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">result</span> <span class="o">=</span> <span class="n">run_rm_correlation</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span> <span class="n">subject</span><span class="o">=</span><span class="s1">&#39;subject&#39;</span><span class="p">,</span> <span class="n">correction</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;fdr&#39;</span><span class="p">,</span> <span class="s1">&#39;indep&#39;</span><span class="p">))</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="analytics_core.analytics.analytics.run_efficient_correlation">
<code class="sig-prename descclassname">analytics_core.analytics.analytics.</code><code class="sig-name descname">run_efficient_correlation</code><span class="sig-paren">(</span><em class="sig-param">data</em>, <em class="sig-param">method='pearson'</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/analytics_core/analytics/analytics.html#run_efficient_correlation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#analytics_core.analytics.analytics.run_efficient_correlation" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates pairwise correlations and returns lower triangle of the matrix with correlation values and p-values.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> – pandas dataframe with samples as index and features as columns (numeric data only).</p></li>
<li><p><strong>method</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – method to use for correlation calculation (‘pearson’, ‘spearman’).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Two numpy arrays: correlation and p-values.</p>
</dd>
</dl>
<p>Example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">result</span> <span class="o">=</span> <span class="n">run_efficient_correlation</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;pearson&#39;</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="analytics_core.analytics.analytics.calculate_paired_ttest">
<code class="sig-prename descclassname">analytics_core.analytics.analytics.</code><code class="sig-name descname">calculate_paired_ttest</code><span class="sig-paren">(</span><em class="sig-param">df</em>, <em class="sig-param">condition1</em>, <em class="sig-param">condition2</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/analytics_core/analytics/analytics.html#calculate_paired_ttest"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#analytics_core.analytics.analytics.calculate_paired_ttest" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates the t-test on RELATED samples belonging to two different groups. For more information visit <a class="reference external" href="https://docs.scipy.org/doc/scipy-0.14.0/reference/generated/scipy.stats.ttest_rel.html">https://docs.scipy.org/doc/scipy-0.14.0/reference/generated/scipy.stats.ttest_rel.html</a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> – pandas dataframe with groups and subjects as rows and protein identifier as column.</p></li>
<li><p><strong>condition1</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – identifier of first group.</p></li>
<li><p><strong>condition2</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – identifier of second group.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Tuple with t-statistics, two-tailed p-value, mean of first group, mean of second group and logfc.</p>
</dd>
</dl>
<p>Example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">result</span> <span class="o">=</span> <span class="n">calculate_paired_ttest</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="s1">&#39;group1&#39;</span><span class="p">,</span> <span class="s1">&#39;group2&#39;</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="analytics_core.analytics.analytics.calculate_ttest_samr">
<code class="sig-prename descclassname">analytics_core.analytics.analytics.</code><code class="sig-name descname">calculate_ttest_samr</code><span class="sig-paren">(</span><em class="sig-param">df</em>, <em class="sig-param">labels</em>, <em class="sig-param">n=2</em>, <em class="sig-param">s0=0</em>, <em class="sig-param">paired=False</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/analytics_core/analytics/analytics.html#calculate_ttest_samr"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#analytics_core.analytics.analytics.calculate_ttest_samr" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates modified T-test using ‘samr’ R package.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> – pandas dataframe with group as columns and protein identifier as rows</p></li>
<li><p><strong>abels</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#list" title="(in Python v3.5)"><em>list</em></a>) – integers reflecting the group each sample belongs to (e.g. group1 = 1, group2 = 2)</p></li>
<li><p><strong>n</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/functions.html#int" title="(in Python v3.5)"><em>int</em></a>) – number of samples</p></li>
<li><p><strong>s0</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/functions.html#float" title="(in Python v3.5)"><em>float</em></a>) – exchangeability factor for denominator of test statistic</p></li>
<li><p><strong>paired</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/functions.html#bool" title="(in Python v3.5)"><em>bool</em></a>) – True if samples are paired</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Pandas dataframe with columns ‘identifier’, ‘group1’, ‘group2’, ‘mean(group1)’, ‘mean(group1)’, ‘log2FC’, ‘FC’, ‘t-statistics’, ‘p-value’.</p>
</dd>
</dl>
<p>Example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">result</span> <span class="o">=</span> <span class="n">calculate_ttest_samr</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">labels</span><span class="p">,</span> <span class="n">n</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">s0</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span> <span class="n">paired</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="analytics_core.analytics.analytics.calculate_ttest">
<code class="sig-prename descclassname">analytics_core.analytics.analytics.</code><code class="sig-name descname">calculate_ttest</code><span class="sig-paren">(</span><em class="sig-param">df</em>, <em class="sig-param">condition1</em>, <em class="sig-param">condition2</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/analytics_core/analytics/analytics.html#calculate_ttest"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#analytics_core.analytics.analytics.calculate_ttest" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates the t-test for the means of independent samples belonging to two different groups. For more information visit <a class="reference external" href="https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ttest_ind.html">https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ttest_ind.html</a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> – pandas dataframe with groups and subjects as rows and protein identifier as column.</p></li>
<li><p><strong>condition1</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – identifier of first group.</p></li>
<li><p><strong>condition2</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – ientifier of second group.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Tuple with t-statistics, two-tailed p-value, mean of first group, mean of second group and logfc.</p>
</dd>
</dl>
<p>Example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">result</span> <span class="o">=</span> <span class="n">calculate_ttest</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="s1">&#39;group1&#39;</span><span class="p">,</span> <span class="s1">&#39;group2&#39;</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="analytics_core.analytics.analytics.calculate_THSD">
<code class="sig-prename descclassname">analytics_core.analytics.analytics.</code><code class="sig-name descname">calculate_THSD</code><span class="sig-paren">(</span><em class="sig-param">df</em>, <em class="sig-param">group='group'</em>, <em class="sig-param">alpha=0.05</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/analytics_core/analytics/analytics.html#calculate_THSD"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#analytics_core.analytics.analytics.calculate_THSD" title="Permalink to this definition">¶</a></dt>
<dd><p>Pairwise Tukey-HSD posthoc test using pingouin stats. For more information visit <a class="reference external" href="https://pingouin-stats.org/generated/pingouin.pairwise_tukey.html">https://pingouin-stats.org/generated/pingouin.pairwise_tukey.html</a></p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> – pandas dataframe with group as rows and protein identifier as column</p></li>
<li><p><strong>group</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – column label containing the within factor</p></li>
<li><p><strong>alpha</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/functions.html#float" title="(in Python v3.5)"><em>float</em></a>) – significance level</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Pandas dataframe.</p>
</dd>
</dl>
<p>Example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">result</span> <span class="o">=</span> <span class="n">calculate_THSD</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">group</span><span class="o">=</span><span class="s1">&#39;group&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.05</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="analytics_core.analytics.analytics.calculate_pairwise_ttest">
<code class="sig-prename descclassname">analytics_core.analytics.analytics.</code><code class="sig-name descname">calculate_pairwise_ttest</code><span class="sig-paren">(</span><em class="sig-param">df</em>, <em class="sig-param">column</em>, <em class="sig-param">subject='subject'</em>, <em class="sig-param">group='group'</em>, <em class="sig-param">correction='none'</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/analytics_core/analytics/analytics.html#calculate_pairwise_ttest"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#analytics_core.analytics.analytics.calculate_pairwise_ttest" title="Permalink to this definition">¶</a></dt>
<dd><p>Performs pairwise t-test using pingouin, as a posthoc test, and calculates fold-changes. For more information visit <a class="reference external" href="https://pingouin-stats.org/generated/pingouin.pairwise_ttests.html">https://pingouin-stats.org/generated/pingouin.pairwise_ttests.html</a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> – pandas dataframe with subject and group as rows and protein identifier as column.</p></li>
<li><p><strong>column</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – column label containing the dependant variable</p></li>
<li><p><strong>subject</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – column label containing subject identifiers</p></li>
<li><p><strong>group</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – column label containing the between factor</p></li>
<li><p><strong>correction</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – method used for testing and adjustment of p-values.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Pandas dataframe with means, standard deviations, test-statistics, degrees of freedom and effect size columns.</p>
</dd>
</dl>
<p>Example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">result</span> <span class="o">=</span> <span class="n">calculate_pairwise_ttest</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="s1">&#39;protein a&#39;</span><span class="p">,</span> <span class="n">subject</span><span class="o">=</span><span class="s1">&#39;subject&#39;</span><span class="p">,</span> <span class="n">group</span><span class="o">=</span><span class="s1">&#39;group&#39;</span><span class="p">,</span> <span class="n">correction</span><span class="o">=</span><span class="s1">&#39;none&#39;</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="analytics_core.analytics.analytics.complement_posthoc">
<code class="sig-prename descclassname">analytics_core.analytics.analytics.</code><code class="sig-name descname">complement_posthoc</code><span class="sig-paren">(</span><em class="sig-param">posthoc</em>, <em class="sig-param">identifier</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/analytics_core/analytics/analytics.html#complement_posthoc"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#analytics_core.analytics.analytics.complement_posthoc" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates fold-changes after posthoc test.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>posthoc</strong> – pandas dataframe from posthoc test. Should have at least columns ‘mean(group1)’ and ‘mean(group2)’.</p></li>
<li><p><strong>identifier</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – feature identifier.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Pandas dataframe with additional columns ‘identifier’, ‘log2FC’ and ‘FC’.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="analytics_core.analytics.analytics.calculate_dabest">
<code class="sig-prename descclassname">analytics_core.analytics.analytics.</code><code class="sig-name descname">calculate_dabest</code><span class="sig-paren">(</span><em class="sig-param">df</em>, <em class="sig-param">idx</em>, <em class="sig-param">x</em>, <em class="sig-param">y</em>, <em class="sig-param">paired=False</em>, <em class="sig-param">id_col=None</em>, <em class="sig-param">test='mean_diff'</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/analytics_core/analytics/analytics.html#calculate_dabest"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#analytics_core.analytics.analytics.calculate_dabest" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> – </p></li>
<li><p><strong>idx</strong> – </p></li>
<li><p><strong>x</strong> – </p></li>
<li><p><strong>y</strong> – </p></li>
<li><p><strong>paired</strong> – </p></li>
<li><p><strong>id_col</strong> – </p></li>
<li><p><strong>test</strong> – </p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="analytics_core.analytics.analytics.calculate_anova_samr">
<code class="sig-prename descclassname">analytics_core.analytics.analytics.</code><code class="sig-name descname">calculate_anova_samr</code><span class="sig-paren">(</span><em class="sig-param">df</em>, <em class="sig-param">labels</em>, <em class="sig-param">s0=0</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/analytics_core/analytics/analytics.html#calculate_anova_samr"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#analytics_core.analytics.analytics.calculate_anova_samr" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates modified one-way ANOVA using ‘samr’ R package.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> – pandas dataframe with group as columns and protein identifier as rows</p></li>
<li><p><strong>labels</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#list" title="(in Python v3.5)"><em>list</em></a>) – integers reflecting the group each sample belongs to (e.g. group1 = 1, group2 = 2, group3 = 3)</p></li>
<li><p><strong>s0</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/functions.html#float" title="(in Python v3.5)"><em>float</em></a>) – exchangeability factor for denominator of test statistic</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Pandas dataframe with protein identifiers and F-statistics.</p>
</dd>
</dl>
<p>Example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">result</span> <span class="o">=</span> <span class="n">calculate_anova_samr</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">labels</span><span class="p">,</span> <span class="n">s0</span><span class="o">=</span><span class="mf">0.1</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="analytics_core.analytics.analytics.calculate_anova">
<code class="sig-prename descclassname">analytics_core.analytics.analytics.</code><code class="sig-name descname">calculate_anova</code><span class="sig-paren">(</span><em class="sig-param">df</em>, <em class="sig-param">group='group'</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/analytics_core/analytics/analytics.html#calculate_anova"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#analytics_core.analytics.analytics.calculate_anova" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates one-way ANOVA using scipy stats.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> – pandas dataframe with group as rows and protein identifier as column</p></li>
<li><p><strong>group</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – column with group identifiers</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Tuple with t-statistics and p-value.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="analytics_core.analytics.analytics.calculate_repeated_measures_anova">
<code class="sig-prename descclassname">analytics_core.analytics.analytics.</code><code class="sig-name descname">calculate_repeated_measures_anova</code><span class="sig-paren">(</span><em class="sig-param">df</em>, <em class="sig-param">column</em>, <em class="sig-param">subject='subject'</em>, <em class="sig-param">group='group'</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/analytics_core/analytics/analytics.html#calculate_repeated_measures_anova"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#analytics_core.analytics.analytics.calculate_repeated_measures_anova" title="Permalink to this definition">¶</a></dt>
<dd><p>One-way and two-way repeated measures ANOVA using pingouin stats.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> – pandas dataframe with samples as rows and protein identifier as column. Data must be in long-format for two-way repeated measures.</p></li>
<li><p><strong>column</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – column label containing the dependant variable</p></li>
<li><p><strong>subject</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – column label containing subject identifiers</p></li>
<li><p><strong>group</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – column label containing the within factor</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Tuple with protein identifier, t-statistics and p-value.</p>
</dd>
</dl>
<p>Example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">result</span> <span class="o">=</span> <span class="n">calculate_repeated_measures_anova</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="s1">&#39;protein a&#39;</span><span class="p">,</span> <span class="n">subject</span><span class="o">=</span><span class="s1">&#39;subject&#39;</span><span class="p">,</span> <span class="n">group</span><span class="o">=</span><span class="s1">&#39;group&#39;</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="analytics_core.analytics.analytics.get_max_permutations">
<code class="sig-prename descclassname">analytics_core.analytics.analytics.</code><code class="sig-name descname">get_max_permutations</code><span class="sig-paren">(</span><em class="sig-param">df</em>, <em class="sig-param">group='group'</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/analytics_core/analytics/analytics.html#get_max_permutations"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#analytics_core.analytics.analytics.get_max_permutations" title="Permalink to this definition">¶</a></dt>
<dd><p>Get maximum number of permutations according to number of samples.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> – pandas dataframe with samples as rows and protein identifiers as columns</p></li>
<li><p><strong>group</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – column with group identifiers</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Maximum number of permutations.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.5/library/functions.html#int" title="(in Python v3.5)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="analytics_core.analytics.analytics.check_is_paired">
<code class="sig-prename descclassname">analytics_core.analytics.analytics.</code><code class="sig-name descname">check_is_paired</code><span class="sig-paren">(</span><em class="sig-param">df</em>, <em class="sig-param">subject</em>, <em class="sig-param">group</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/analytics_core/analytics/analytics.html#check_is_paired"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#analytics_core.analytics.analytics.check_is_paired" title="Permalink to this definition">¶</a></dt>
<dd><p>Check if samples are paired.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> – pandas dataframe with samples as rows and protein identifiers as columns (with additional columns ‘group’, ‘sample’ and ‘subject’).</p></li>
<li><p><strong>subject</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – column with subject identifiers</p></li>
<li><p><strong>group</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – column with group identifiers</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>True if paired samples.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.5/library/functions.html#bool" title="(in Python v3.5)">bool</a></p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="analytics_core.analytics.analytics.run_dabest">
<code class="sig-prename descclassname">analytics_core.analytics.analytics.</code><code class="sig-name descname">run_dabest</code><span class="sig-paren">(</span><em class="sig-param">df, drop_cols=['sample'], subject='subject', group='group', test='mean_diff'</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/analytics_core/analytics/analytics.html#run_dabest"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#analytics_core.analytics.analytics.run_dabest" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> – </p></li>
<li><p><strong>drop_cols</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#list" title="(in Python v3.5)"><em>list</em></a>) – </p></li>
<li><p><strong>subject</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – </p></li>
<li><p><strong>group</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – </p></li>
<li><p><strong>test</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – </p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Pandas dataframe</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="analytics_core.analytics.analytics.run_anova">
<code class="sig-prename descclassname">analytics_core.analytics.analytics.</code><code class="sig-name descname">run_anova</code><span class="sig-paren">(</span><em class="sig-param">df, alpha=0.05, drop_cols=['sample', 'subject'], subject='subject', group='group', permutations=50</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/analytics_core/analytics/analytics.html#run_anova"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#analytics_core.analytics.analytics.run_anova" title="Permalink to this definition">¶</a></dt>
<dd><p>Performs statistical test for each protein in a dataset.
Checks what type of data is the input (paired, unpaired or repeated measurements) and performs posthoc tests for multiclass data.
Multiple hypothesis correction uses permutation-based if permutations&gt;0 and Benjamini/Hochberg if permutations=0.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> – pandas dataframe with samples as rows and protein identifiers as columns (with additional columns ‘group’, ‘sample’ and ‘subject’).</p></li>
<li><p><strong>subject</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – column with subject identifiers</p></li>
<li><p><strong>group</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – column with group identifiers</p></li>
<li><p><strong>drop_cols</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#list" title="(in Python v3.5)"><em>list</em></a>) – column labels to be dropped from the dataframe</p></li>
<li><p><strong>alpha</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/functions.html#float" title="(in Python v3.5)"><em>float</em></a>) – error rate for multiple hypothesis correction</p></li>
<li><p><strong>permutations</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/functions.html#int" title="(in Python v3.5)"><em>int</em></a>) – number of permutations used to estimate false discovery rates.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Pandas dataframe with columns ‘identifier’, ‘group1’, ‘group2’, ‘mean(group1)’, ‘mean(group2)’, ‘Log2FC’, ‘std_error’, ‘tail’, ‘t-statistics’, ‘padj_THSD’, ‘effsize’, ‘efftype’, ‘FC’, ‘rejected’, ‘F-statistics’, ‘p-value’, ‘correction’, ‘-log10 p-value’, and ‘method’.</p>
</dd>
</dl>
<p>Example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">result</span> <span class="o">=</span> <span class="n">run_anova</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span> <span class="n">drop_cols</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;sample&quot;</span><span class="p">,</span><span class="s1">&#39;subject&#39;</span><span class="p">],</span> <span class="n">subject</span><span class="o">=</span><span class="s1">&#39;subject&#39;</span><span class="p">,</span> <span class="n">group</span><span class="o">=</span><span class="s1">&#39;group&#39;</span><span class="p">,</span> <span class="n">permutations</span><span class="o">=</span><span class="mi">50</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="analytics_core.analytics.analytics.run_repeated_measurements_anova">
<code class="sig-prename descclassname">analytics_core.analytics.analytics.</code><code class="sig-name descname">run_repeated_measurements_anova</code><span class="sig-paren">(</span><em class="sig-param">df, alpha=0.05, drop_cols=['sample'], subject='subject', group='group', permutations=50</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/analytics_core/analytics/analytics.html#run_repeated_measurements_anova"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#analytics_core.analytics.analytics.run_repeated_measurements_anova" title="Permalink to this definition">¶</a></dt>
<dd><p>Performs repeated measurements anova and pairwise posthoc tests for each protein in dataframe.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> – pandas dataframe with samples as rows and protein identifiers as columns (with additional columns ‘group’, ‘sample’ and ‘subject’).</p></li>
<li><p><strong>subject</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – column with subject identifiers</p></li>
<li><p><strong>group</strong> (<em>srt</em>) – column with group identifiers</p></li>
<li><p><strong>drop_cols</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#list" title="(in Python v3.5)"><em>list</em></a>) – column labels to be dropped from the dataframe</p></li>
<li><p><strong>alpha</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/functions.html#float" title="(in Python v3.5)"><em>float</em></a>) – error rate for multiple hypothesis correction</p></li>
<li><p><strong>permutations</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/functions.html#int" title="(in Python v3.5)"><em>int</em></a>) – number of permutations used to estimate false discovery rates</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Pandas dataframe</p>
</dd>
</dl>
<p>Example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">result</span> <span class="o">=</span> <span class="n">run_repeated_measurements_anova</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span> <span class="n">drop_cols</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;sample&#39;</span><span class="p">],</span> <span class="n">subject</span><span class="o">=</span><span class="s1">&#39;subject&#39;</span><span class="p">,</span> <span class="n">group</span><span class="o">=</span><span class="s1">&#39;group&#39;</span><span class="p">,</span> <span class="n">permutations</span><span class="o">=</span><span class="mi">50</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="analytics_core.analytics.analytics.format_anova_table">
<code class="sig-prename descclassname">analytics_core.analytics.analytics.</code><code class="sig-name descname">format_anova_table</code><span class="sig-paren">(</span><em class="sig-param">df</em>, <em class="sig-param">aov_results</em>, <em class="sig-param">pairwise_results</em>, <em class="sig-param">group</em>, <em class="sig-param">permutations</em>, <em class="sig-param">alpha</em>, <em class="sig-param">max_permutations</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/analytics_core/analytics/analytics.html#format_anova_table"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#analytics_core.analytics.analytics.format_anova_table" title="Permalink to this definition">¶</a></dt>
<dd><p>Performs p-value correction (permutation-based and FDR) and converts pandas dataframe into final format.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> – pandas dataframe with samples as rows and protein identifiers as columns (with additional columns ‘group’, ‘sample’ and ‘subject’).</p></li>
<li><p><strong>aov_results</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#list" title="(in Python v3.5)"><em>list</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#tuple" title="(in Python v3.5)"><em>tuple</em></a><em>]</em>) – list of tuples with anova results (one tuple per feature).</p></li>
<li><p><strong>pairwise_results</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#list" title="(in Python v3.5)"><em>list</em></a><em>[</em><em>dataframes</em><em>]</em>) – list of pandas dataframes with posthoc tests results</p></li>
<li><p><strong>group</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – column with group identifiers</p></li>
<li><p><strong>alpha</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/functions.html#float" title="(in Python v3.5)"><em>float</em></a>) – error rate for multiple hypothesis correction</p></li>
<li><p><strong>permutations</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/functions.html#int" title="(in Python v3.5)"><em>int</em></a>) – number of permutations used to estimate false discovery rates</p></li>
<li><p><strong>max_permutations</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/functions.html#int" title="(in Python v3.5)"><em>int</em></a>) – maximum number of permutations according to number of samples</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Pandas dataframe</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="analytics_core.analytics.analytics.run_ttest">
<code class="sig-prename descclassname">analytics_core.analytics.analytics.</code><code class="sig-name descname">run_ttest</code><span class="sig-paren">(</span><em class="sig-param">df, condition1, condition2, alpha=0.05, drop_cols=['sample'], subject='subject', group='group', paired=False, correction='indep', permutations=50</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/analytics_core/analytics/analytics.html#run_ttest"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#analytics_core.analytics.analytics.run_ttest" title="Permalink to this definition">¶</a></dt>
<dd><p>Runs t-test (paired/unpaired) for each protein in dataset and performs permutation-based (if permutations&gt;0) or Benjamini/Hochberg (if permutations=0) multiple hypothesis correction.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> – pandas dataframe with samples as rows and protein identifiers as columns (with additional columns ‘group’, ‘sample’ and ‘subject’).</p></li>
<li><p><strong>condition1</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – first of two conditions of the independent variable</p></li>
<li><p><strong>condition2</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – second of two conditions of the independent variable</p></li>
<li><p><strong>subject</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – column with subject identifiers</p></li>
<li><p><strong>group</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – column with group identifiers (independent variable)</p></li>
<li><p><strong>drop_cols</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#list" title="(in Python v3.5)"><em>list</em></a>) – column labels to be dropped from the dataframe</p></li>
<li><p><strong>paired</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/functions.html#bool" title="(in Python v3.5)"><em>bool</em></a>) – paired or unpaired samples</p></li>
<li><p><strong>correction</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – method of pvalue correction for false discovery rate (‘indep’, ‘negcorr’)</p></li>
<li><p><strong>alpha</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/functions.html#float" title="(in Python v3.5)"><em>float</em></a>) – error rate for multiple hypothesis correction</p></li>
<li><p><strong>permutations</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/functions.html#int" title="(in Python v3.5)"><em>int</em></a>) – number of permutations used to estimate false discovery rates.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Pandas dataframe with columns ‘identifier’, ‘group1’, ‘group2’, ‘mean(group1)’, ‘mean(group2)’, ‘Log2FC’, ‘FC’, ‘rejected’, ‘T-statistics’, ‘p-value’, ‘correction’, ‘-log10 p-value’, and ‘method’.</p>
</dd>
</dl>
<p>Example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">result</span> <span class="o">=</span> <span class="n">run_ttest</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">condition1</span><span class="o">=</span><span class="s1">&#39;group1&#39;</span><span class="p">,</span> <span class="n">condition2</span><span class="o">=</span><span class="s1">&#39;group2&#39;</span><span class="p">,</span> <span class="n">alpha</span> <span class="o">=</span> <span class="mf">0.05</span><span class="p">,</span> <span class="n">drop_cols</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;sample&#39;</span><span class="p">],</span> <span class="n">subject</span><span class="o">=</span><span class="s1">&#39;subject&#39;</span><span class="p">,</span> <span class="n">group</span><span class="o">=</span><span class="s1">&#39;group&#39;</span><span class="p">,</span> <span class="n">paired</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">correction</span><span class="o">=</span><span class="s1">&#39;indep&#39;</span><span class="p">,</span> <span class="n">permutations</span><span class="o">=</span><span class="mi">50</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="analytics_core.analytics.analytics.run_samr">
<code class="sig-prename descclassname">analytics_core.analytics.analytics.</code><code class="sig-name descname">run_samr</code><span class="sig-paren">(</span><em class="sig-param">df, subject='subject', group='group', drop_cols=['subject', 'sample'], alpha=0.05, s0=1, permutations=250</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/analytics_core/analytics/analytics.html#run_samr"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#analytics_core.analytics.analytics.run_samr" title="Permalink to this definition">¶</a></dt>
<dd><p>Python adaptation of the ‘samr’ R package for statistical tests with permutation-based correction and s0 parameter.
For more information visit <a class="reference external" href="https://cran.r-project.org/web/packages/samr/samr.pdf">https://cran.r-project.org/web/packages/samr/samr.pdf</a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> – pandas dataframe with samples as rows and protein identifiers as columns (with additional columns ‘group’, ‘sample’ and ‘subject’).</p></li>
<li><p><strong>subject</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – column with subject identifiers</p></li>
<li><p><strong>group</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – column with group identifiers</p></li>
<li><p><strong>drop_cols</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#list" title="(in Python v3.5)"><em>list</em></a>) – columnlabels to be dropped from the dataframe</p></li>
<li><p><strong>alpha</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/functions.html#float" title="(in Python v3.5)"><em>float</em></a>) – error rate for multiple hypothesis correction</p></li>
<li><p><strong>s0</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/functions.html#float" title="(in Python v3.5)"><em>float</em></a>) – exchangeability factor for denominator of test statistic</p></li>
<li><p><strong>permutations</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/functions.html#int" title="(in Python v3.5)"><em>int</em></a>) – number of permutations used to estimate false discovery rates. If number of permutations is equal to zero, the function will run anova with FDR Benjamini/Hochberg correction.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Pandas dataframe with columns ‘identifier’, ‘group1’, ‘group2’, ‘mean(group1)’, ‘mean(group2)’, ‘Log2FC’, ‘FC’, ‘T-statistics’, ‘p-value’, ‘padj’, ‘correction’, ‘-log10 p-value’, ‘rejected’ and ‘method’</p>
</dd>
</dl>
<p>Example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">result</span> <span class="o">=</span> <span class="n">run_samr</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">subject</span><span class="o">=</span><span class="s1">&#39;subject&#39;</span><span class="p">,</span> <span class="n">group</span><span class="o">=</span><span class="s1">&#39;group&#39;</span><span class="p">,</span> <span class="n">drop_cols</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;subject&#39;</span><span class="p">,</span> <span class="s1">&#39;sample&#39;</span><span class="p">],</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span> <span class="n">s0</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">permutations</span><span class="o">=</span><span class="mi">250</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="analytics_core.analytics.analytics.run_fisher">
<code class="sig-prename descclassname">analytics_core.analytics.analytics.</code><code class="sig-name descname">run_fisher</code><span class="sig-paren">(</span><em class="sig-param">group1</em>, <em class="sig-param">group2</em>, <em class="sig-param">alternative='two-sided'</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/analytics_core/analytics/analytics.html#run_fisher"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#analytics_core.analytics.analytics.run_fisher" title="Permalink to this definition">¶</a></dt>
<dd><p>annotated   not-annotated
group1      a               b
group2      c               d
————————————</p>
<p>group1 = [a, b]
group2 = [c, d]</p>
<p>odds, pvalue = stats.fisher_exact([[a, b], [c, d]])</p>
</dd></dl>

<dl class="function">
<dt id="analytics_core.analytics.analytics.run_regulation_enrichment">
<code class="sig-prename descclassname">analytics_core.analytics.analytics.</code><code class="sig-name descname">run_regulation_enrichment</code><span class="sig-paren">(</span><em class="sig-param">regulation_data, annotation, identifier='identifier', groups=['group1', 'group2'], annotation_col='annotation', reject_col='rejected', group_col='group', method='fisher'</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/analytics_core/analytics/analytics.html#run_regulation_enrichment"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#analytics_core.analytics.analytics.run_regulation_enrichment" title="Permalink to this definition">¶</a></dt>
<dd><p>This function runs a simple enrichment analysis for significantly regulated features in a dataset.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>regulation_data</strong> – pandas dataframe resulting from differential regulation analysis.</p></li>
<li><p><strong>annotation</strong> – pandas dataframe with annotations for features (columns: ‘annotation’, ‘identifier’ (feature identifiers), and ‘source’).</p></li>
<li><p><strong>identifier</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – name of the column from annotation containing feature identifiers.</p></li>
<li><p><strong>groups</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#list" title="(in Python v3.5)"><em>list</em></a>) – column names from regulation_data containing group identifiers.</p></li>
<li><p><strong>annotation_col</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – name of the column from annotation containing annotation terms.</p></li>
<li><p><strong>reject_col</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – name of the column from regulatio_data containing boolean for rejected null hypothesis.</p></li>
<li><p><strong>group_col</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – column name for new column in annotation dataframe determining if feature belongs to foreground or background.</p></li>
<li><p><strong>method</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – method used to compute enrichment (only ‘fisher’ is supported currently).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Pandas dataframe with columns: ‘terms’, ‘identifiers’, ‘foreground’, ‘background’, ‘pvalue’, ‘padj’ and ‘rejected’.</p>
</dd>
</dl>
<p>Example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">result</span> <span class="o">=</span> <span class="n">run_regulation_enrichment</span><span class="p">(</span><span class="n">regulation_data</span><span class="p">,</span> <span class="n">annotation</span><span class="p">,</span> <span class="n">identifier</span><span class="o">=</span><span class="s1">&#39;identifier&#39;</span><span class="p">,</span> <span class="n">groups</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;group1&#39;</span><span class="p">,</span> <span class="s1">&#39;group2&#39;</span><span class="p">],</span> <span class="n">annotation_col</span><span class="o">=</span><span class="s1">&#39;annotation&#39;</span><span class="p">,</span> <span class="n">reject_col</span><span class="o">=</span><span class="s1">&#39;rejected&#39;</span><span class="p">,</span> <span class="n">group_col</span><span class="o">=</span><span class="s1">&#39;group&#39;</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;fisher&#39;</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="analytics_core.analytics.analytics.run_enrichment">
<code class="sig-prename descclassname">analytics_core.analytics.analytics.</code><code class="sig-name descname">run_enrichment</code><span class="sig-paren">(</span><em class="sig-param">data</em>, <em class="sig-param">foreground</em>, <em class="sig-param">background</em>, <em class="sig-param">foreground_pop</em>, <em class="sig-param">background_pop</em>, <em class="sig-param">annotation_col='annotation'</em>, <em class="sig-param">group_col='group'</em>, <em class="sig-param">identifier_col='identifier'</em>, <em class="sig-param">method='fisher'</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/analytics_core/analytics/analytics.html#run_enrichment"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#analytics_core.analytics.analytics.run_enrichment" title="Permalink to this definition">¶</a></dt>
<dd><p>Computes enrichment of the foreground relative to a given backgroung, using Fisher’s exact test, and corrects for multiple hypothesis testing.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> – pandas dataframe with annotations for dataset features (columns: ‘annotation’, ‘identifier’, ‘source’, ‘group’).</p></li>
<li><p><strong>foreground</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – group identifier of features that belong to the foreground.</p></li>
<li><p><strong>background</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – group identifier of features that belong to the background.</p></li>
<li><p><strong>foreground_pop</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/functions.html#int" title="(in Python v3.5)"><em>int</em></a>) – number of features in the foreground population.</p></li>
<li><p><strong>background_pop</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/functions.html#int" title="(in Python v3.5)"><em>int</em></a>) – number of features in the background population.</p></li>
<li><p><strong>annotation_col</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – name of the column containing annotation terms.</p></li>
<li><p><strong>group_col</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – name of column containing the group identifiers.</p></li>
<li><p><strong>identifier_col</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – name of column containing dependent variables identifiers.</p></li>
<li><p><strong>method</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – method used to compute enrichment (only ‘fisher’ is supported currently).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Pandas dataframe with annotation terms, features, number of foregroung/background features in each term, p-values and corrected p-values (columns: ‘terms’, ‘identifiers’, ‘foreground’, ‘background’, ‘pvalue’, ‘padj’ and ‘rejected’).</p>
</dd>
</dl>
<p>Example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">result</span> <span class="o">=</span> <span class="n">run_enrichment</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">foreground</span><span class="o">=</span><span class="s1">&#39;foreground&#39;</span><span class="p">,</span> <span class="n">background</span><span class="o">=</span><span class="s1">&#39;background&#39;</span><span class="p">,</span> <span class="n">foreground_pop</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="n">foreground_list</span><span class="p">),</span> <span class="n">background_pop</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="n">background_list</span><span class="p">),</span> <span class="n">annotation_col</span><span class="o">=</span><span class="s1">&#39;annotation&#39;</span><span class="p">,</span> <span class="n">group_col</span><span class="o">=</span><span class="s1">&#39;group&#39;</span><span class="p">,</span> <span class="n">identifier_col</span><span class="o">=</span><span class="s1">&#39;identifier&#39;</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;fisher&#39;</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="analytics_core.analytics.analytics.calculate_fold_change">
<code class="sig-prename descclassname">analytics_core.analytics.analytics.</code><code class="sig-name descname">calculate_fold_change</code><span class="sig-paren">(</span><em class="sig-param">df</em>, <em class="sig-param">condition1</em>, <em class="sig-param">condition2</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/analytics_core/analytics/analytics.html#calculate_fold_change"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#analytics_core.analytics.analytics.calculate_fold_change" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates fold-changes between two groups for all proteins in a dataframe.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> – pandas dataframe with samples as rows and protein identifiers as columns.</p></li>
<li><p><strong>condition1</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – identifier of first group.</p></li>
<li><p><strong>condition2</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – identifier of second group.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Numpy array.</p>
</dd>
</dl>
<p>Example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">result</span> <span class="o">=</span> <span class="n">calculate_fold_change</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="s1">&#39;group1&#39;</span><span class="p">,</span> <span class="s1">&#39;group2&#39;</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="analytics_core.analytics.analytics.cohen_d">
<code class="sig-prename descclassname">analytics_core.analytics.analytics.</code><code class="sig-name descname">cohen_d</code><span class="sig-paren">(</span><em class="sig-param">df</em>, <em class="sig-param">condition1</em>, <em class="sig-param">condition2</em>, <em class="sig-param">ddof=0</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/analytics_core/analytics/analytics.html#cohen_d"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#analytics_core.analytics.analytics.cohen_d" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates Cohen’s d effect size based on the distance between two means, measured in standard deviations.
For more information visit <a class="reference external" href="https://docs.scipy.org/doc/numpy/reference/generated/numpy.nanstd.html">https://docs.scipy.org/doc/numpy/reference/generated/numpy.nanstd.html</a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> – pandas dataframe with samples as rows and protein identifiers as columns.</p></li>
<li><p><strong>condition1</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – identifier of first group.</p></li>
<li><p><strong>condition2</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – identifier of second group.</p></li>
<li><p><strong>ddof</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/functions.html#int" title="(in Python v3.5)"><em>int</em></a>) – means Delta Degrees of Freedom.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Numpy array.</p>
</dd>
</dl>
<p>Example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">result</span> <span class="o">=</span> <span class="n">cohen_d</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="s1">&#39;group1&#39;</span><span class="p">,</span> <span class="s1">&#39;group2&#39;</span><span class="p">,</span> <span class="n">ddof</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="analytics_core.analytics.analytics.hedges_g">
<code class="sig-prename descclassname">analytics_core.analytics.analytics.</code><code class="sig-name descname">hedges_g</code><span class="sig-paren">(</span><em class="sig-param">df</em>, <em class="sig-param">condition1</em>, <em class="sig-param">condition2</em>, <em class="sig-param">ddof=0</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/analytics_core/analytics/analytics.html#hedges_g"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#analytics_core.analytics.analytics.hedges_g" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates Hedges’ g effect size (more accurate for sample sizes below 20 than Cohen’s d).
For more information visit <a class="reference external" href="https://docs.scipy.org/doc/numpy/reference/generated/numpy.nanstd.html">https://docs.scipy.org/doc/numpy/reference/generated/numpy.nanstd.html</a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> – pandas dataframe with samples as rows and protein identifiers as columns.</p></li>
<li><p><strong>condition1</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – identifier of first group.</p></li>
<li><p><strong>condition2</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – identifier of second group.</p></li>
<li><p><strong>ddof</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/functions.html#int" title="(in Python v3.5)"><em>int</em></a>) – means Delta Degrees of Freedom.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Numpy array.</p>
</dd>
</dl>
<p>Example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">result</span> <span class="o">=</span> <span class="n">hedges_g</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="s1">&#39;group1&#39;</span><span class="p">,</span> <span class="s1">&#39;group2&#39;</span><span class="p">,</span> <span class="n">ddof</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="analytics_core.analytics.analytics.run_mapper">
<code class="sig-prename descclassname">analytics_core.analytics.analytics.</code><code class="sig-name descname">run_mapper</code><span class="sig-paren">(</span><em class="sig-param">data, lenses=['l2norm'], n_cubes=15, overlap=0.5, n_clusters=3, linkage='complete', affinity='correlation'</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/analytics_core/analytics/analytics.html#run_mapper"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#analytics_core.analytics.analytics.run_mapper" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> – </p></li>
<li><p><strong>lenses</strong> – </p></li>
<li><p><strong>n_cubes</strong> – </p></li>
<li><p><strong>overlap</strong> – </p></li>
<li><p><strong>n_clusters</strong> – </p></li>
<li><p><strong>linkage</strong> – </p></li>
<li><p><strong>affinity</strong> – </p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="analytics_core.analytics.analytics.run_WGCNA">
<code class="sig-prename descclassname">analytics_core.analytics.analytics.</code><code class="sig-name descname">run_WGCNA</code><span class="sig-paren">(</span><em class="sig-param">data</em>, <em class="sig-param">drop_cols_exp</em>, <em class="sig-param">drop_cols_cli</em>, <em class="sig-param">RsquaredCut=0.8</em>, <em class="sig-param">networkType='unsigned'</em>, <em class="sig-param">minModuleSize=30</em>, <em class="sig-param">deepSplit=2</em>, <em class="sig-param">pamRespectsDendro=False</em>, <em class="sig-param">merge_modules=True</em>, <em class="sig-param">MEDissThres=0.25</em>, <em class="sig-param">verbose=0</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/analytics_core/analytics/analytics.html#run_WGCNA"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#analytics_core.analytics.analytics.run_WGCNA" title="Permalink to this definition">¶</a></dt>
<dd><p>Runs an automated weighted gene co-expression network analysis (WGCNA), using input proteomics/transcriptomics/genomics and clinical variables data.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#dict" title="(in Python v3.5)"><em>dict</em></a>) – dictionary of pandas dataframes with processed clinical and experimental datasets</p></li>
<li><p><strong>drop_cols_exp</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#list" title="(in Python v3.5)"><em>list</em></a>) – column names to be removed from the experimental dataset.</p></li>
<li><p><strong>drop_cols_cli</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#list" title="(in Python v3.5)"><em>list</em></a>) – column names to be removed from the clinical dataset.</p></li>
<li><p><strong>RsquaredCut</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/functions.html#float" title="(in Python v3.5)"><em>float</em></a>) – desired minimum scale free topology fitting index R^2.</p></li>
<li><p><strong>networkType</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – network type (‘unsigned’, ‘signed’, ‘signed hybrid’, ‘distance’).</p></li>
<li><p><strong>minModuleSize</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/functions.html#int" title="(in Python v3.5)"><em>int</em></a>) – minimum module size.</p></li>
<li><p><strong>deepSplit</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/functions.html#int" title="(in Python v3.5)"><em>int</em></a>) – provides a rough control over sensitivity to cluster splitting, the higher the value (with ‘hybrid’ method) or if True (with ‘tree’ method), the more and smaller modules.</p></li>
<li><p><strong>pamRespectsDendro</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/functions.html#bool" title="(in Python v3.5)"><em>bool</em></a>) – only used for method ‘hybrid’. Objects and small modules will only be assigned to modules that belong to the same branch in the dendrogram structure.</p></li>
<li><p><strong>merge_modules</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/functions.html#bool" title="(in Python v3.5)"><em>bool</em></a>) – if True, very similar modules are merged.</p></li>
<li><p><strong>MEDissThres</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/functions.html#float" title="(in Python v3.5)"><em>float</em></a>) – maximum dissimilarity (i.e., 1-correlation) that qualifies modules for merging.</p></li>
<li><p><strong>verbose</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/functions.html#int" title="(in Python v3.5)"><em>int</em></a>) – integer level of verbosity. Zero means silent, higher values make the output progressively more and more verbose.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Tuple with multiple pandas dataframes.</p>
</dd>
</dl>
<p>Example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">result</span> <span class="o">=</span> <span class="n">run_WGCNA</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">drop_cols_exp</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;subject&#39;</span><span class="p">,</span> <span class="s1">&#39;sample&#39;</span><span class="p">,</span> <span class="s1">&#39;group&#39;</span><span class="p">,</span> <span class="s1">&#39;index&#39;</span><span class="p">],</span> <span class="n">drop_cols_cli</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;subject&#39;</span><span class="p">,</span> <span class="s1">&#39;biological_sample&#39;</span><span class="p">,</span> <span class="s1">&#39;group&#39;</span><span class="p">,</span> <span class="s1">&#39;index&#39;</span><span class="p">],</span> <span class="n">RsquaredCut</span><span class="o">=</span><span class="mf">0.8</span><span class="p">,</span> <span class="n">networkType</span><span class="o">=</span><span class="s1">&#39;unsigned&#39;</span><span class="p">,</span> <span class="n">minModuleSize</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">deepSplit</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">pamRespectsDendro</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">merge_modules</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">MEDissThres</span><span class="o">=</span><span class="mf">0.25</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="analytics_core.analytics.analytics.most_central_edge">
<code class="sig-prename descclassname">analytics_core.analytics.analytics.</code><code class="sig-name descname">most_central_edge</code><span class="sig-paren">(</span><em class="sig-param">G</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/analytics_core/analytics/analytics.html#most_central_edge"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#analytics_core.analytics.analytics.most_central_edge" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute the eigenvector centrality for the graph G, and finds the highest value.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>G</strong> (<em>graph</em>) – networkx graph</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Highest eigenvector centrality value.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.5/library/functions.html#float" title="(in Python v3.5)">float</a></p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="analytics_core.analytics.analytics.get_louvain_partitions">
<code class="sig-prename descclassname">analytics_core.analytics.analytics.</code><code class="sig-name descname">get_louvain_partitions</code><span class="sig-paren">(</span><em class="sig-param">G</em>, <em class="sig-param">weight</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/analytics_core/analytics/analytics.html#get_louvain_partitions"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#analytics_core.analytics.analytics.get_louvain_partitions" title="Permalink to this definition">¶</a></dt>
<dd><p>Computes the partition of the graph nodes which maximises the modularity (or try..) using the Louvain heuristices. For more information visit <a class="reference external" href="https://python-louvain.readthedocs.io/en/latest/api.html">https://python-louvain.readthedocs.io/en/latest/api.html</a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>G</strong> (<em>graph</em>) – networkx graph which is decomposed.</p></li>
<li><p><strong>weight</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – the key in graph to use as weight.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The partition, with communities numbered from 0 to number of communities.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#dict" title="(in Python v3.5)">dict</a></p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="analytics_core.analytics.analytics.get_network_communities">
<code class="sig-prename descclassname">analytics_core.analytics.analytics.</code><code class="sig-name descname">get_network_communities</code><span class="sig-paren">(</span><em class="sig-param">graph</em>, <em class="sig-param">args</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/analytics_core/analytics/analytics.html#get_network_communities"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#analytics_core.analytics.analytics.get_network_communities" title="Permalink to this definition">¶</a></dt>
<dd><p>Finds communities in a graph using different methods. For more information on the methods visit:</p>
<blockquote>
<div><ul class="simple">
<li><p><a class="reference external" href="https://networkx.github.io/documentation/latest/reference/algorithms/generated/networkx.algorithms.community.modularity_max.greedy_modularity_communities.html">https://networkx.github.io/documentation/latest/reference/algorithms/generated/networkx.algorithms.community.modularity_max.greedy_modularity_communities.html</a></p></li>
<li><p><a class="reference external" href="https://networkx.github.io/documentation/networkx-2.0/reference/algorithms/generated/networkx.algorithms.community.asyn_lpa.asyn_lpa_communities.html">https://networkx.github.io/documentation/networkx-2.0/reference/algorithms/generated/networkx.algorithms.community.asyn_lpa.asyn_lpa_communities.html</a></p></li>
<li><p><a class="reference external" href="https://networkx.github.io/documentation/latest/reference/algorithms/generated/networkx.algorithms.community.centrality.girvan_newman.html">https://networkx.github.io/documentation/latest/reference/algorithms/generated/networkx.algorithms.community.centrality.girvan_newman.html</a></p></li>
<li><p><a class="reference external" href="https://networkx.github.io/documentation/latest/reference/generated/networkx.convert_matrix.to_pandas_adjacency.html">https://networkx.github.io/documentation/latest/reference/generated/networkx.convert_matrix.to_pandas_adjacency.html</a></p></li>
</ul>
</div></blockquote>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>graph</strong> (<em>graph</em>) – networkx graph</p></li>
<li><p><strong>args</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#dict" title="(in Python v3.5)"><em>dict</em></a>) – config file arguments</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Dictionary of nodes and which community they belong to (from 0 to number of communities).</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="analytics_core.analytics.analytics.get_publications_abstracts">
<code class="sig-prename descclassname">analytics_core.analytics.analytics.</code><code class="sig-name descname">get_publications_abstracts</code><span class="sig-paren">(</span><em class="sig-param">data, publication_col='publication', join_by=['publication', 'Proteins', 'Diseases'], index='PMID'</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/analytics_core/analytics/analytics.html#get_publications_abstracts"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#analytics_core.analytics.analytics.get_publications_abstracts" title="Permalink to this definition">¶</a></dt>
<dd><p>Accesses NCBI PubMed over the WWW and retrieves the abstracts corresponding to a list of one or more PubMed IDs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> – pandas dataframe of diseases and publications linked to a list of proteins (columns: ‘Diseases’, ‘Proteins’, ‘linkout’ and ‘publication’).</p></li>
<li><p><strong>publication_col</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – column label containing PubMed ids.</p></li>
<li><p><strong>join_by</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#list" title="(in Python v3.5)"><em>list</em></a>) – column labels to be kept from the input dataframe.</p></li>
<li><p><strong>index</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – column label containing PubMed ids from the NCBI retrieved data.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Pandas dataframe with publication information and columns ‘PMID’, ‘abstract’, ‘authors’, ‘date’, ‘journal’, ‘keywords’, ‘title’, ‘url’, ‘Proteins’ and ‘Diseases’.</p>
</dd>
</dl>
<p>Example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">result</span> <span class="o">=</span> <span class="n">get_publications_abstracts</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">publication_col</span><span class="o">=</span><span class="s1">&#39;publication&#39;</span><span class="p">,</span> <span class="n">join_by</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;publication&#39;</span><span class="p">,</span><span class="s1">&#39;Proteins&#39;</span><span class="p">,</span><span class="s1">&#39;Diseases&#39;</span><span class="p">],</span> <span class="n">index</span><span class="o">=</span><span class="s1">&#39;PMID&#39;</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="analytics_core.analytics.analytics.eta_squared">
<code class="sig-prename descclassname">analytics_core.analytics.analytics.</code><code class="sig-name descname">eta_squared</code><span class="sig-paren">(</span><em class="sig-param">aov</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/analytics_core/analytics/analytics.html#eta_squared"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#analytics_core.analytics.analytics.eta_squared" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates the effect size using Eta-squared.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>aov</strong> – pandas dataframe with anova results from statsmodels.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Pandas dataframe with additional Eta-squared column.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="analytics_core.analytics.analytics.omega_squared">
<code class="sig-prename descclassname">analytics_core.analytics.analytics.</code><code class="sig-name descname">omega_squared</code><span class="sig-paren">(</span><em class="sig-param">aov</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/analytics_core/analytics/analytics.html#omega_squared"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#analytics_core.analytics.analytics.omega_squared" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates the effect size using Omega-squared.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>aov</strong> – pandas dataframe with anova results from statsmodels.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Pandas dataframe with additional Omega-squared column.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="analytics_core.analytics.analytics.run_two_way_anova">
<code class="sig-prename descclassname">analytics_core.analytics.analytics.</code><code class="sig-name descname">run_two_way_anova</code><span class="sig-paren">(</span><em class="sig-param">df, drop_cols=['sample'], subject='subject', group=['group', 'secondary_group']</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/analytics_core/analytics/analytics.html#run_two_way_anova"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#analytics_core.analytics.analytics.run_two_way_anova" title="Permalink to this definition">¶</a></dt>
<dd><p>Run a 2-way ANOVA when data[‘secondary_group’] is not empty</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> – processed pandas dataframe with samples as rows, and proteins and groups as columns.</p></li>
<li><p><strong>drop_cols</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#list" title="(in Python v3.5)"><em>list</em></a>) – column names to drop from dataframe</p></li>
<li><p><strong>subject</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – column name containing subject identifiers.</p></li>
<li><p><strong>group</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#list" title="(in Python v3.5)"><em>list</em></a>) – column names corresponding to independent variable groups</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Two dataframes, anova results and residuals.</p>
</dd>
</dl>
<p>Example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">result</span> <span class="o">=</span> <span class="n">run_two_way_anova</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">drop_cols</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;sample&#39;</span><span class="p">],</span> <span class="n">subject</span><span class="o">=</span><span class="s1">&#39;subject&#39;</span><span class="p">,</span> <span class="n">group</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;group&#39;</span><span class="p">,</span> <span class="s1">&#39;secondary_group&#39;</span><span class="p">])</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="analytics_core.analytics.analytics.run_snf">
<code class="sig-prename descclassname">analytics_core.analytics.analytics.</code><code class="sig-name descname">run_snf</code><span class="sig-paren">(</span><em class="sig-param">df_dict</em>, <em class="sig-param">clusters</em>, <em class="sig-param">distance_metric</em>, <em class="sig-param">K_affinity</em>, <em class="sig-param">mu_affinity</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/analytics_core/analytics/analytics.html#run_snf"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#analytics_core.analytics.analytics.run_snf" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df_dict</strong> – </p></li>
<li><p><strong>clusters</strong> – </p></li>
</ul>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-analytics_core.analytics.wgcnaAnalysis">
<span id="wgcna-analytics"></span><h2>WGCNA analytics<a class="headerlink" href="#module-analytics_core.analytics.wgcnaAnalysis" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="analytics_core.analytics.wgcnaAnalysis.get_data">
<code class="sig-prename descclassname">analytics_core.analytics.wgcnaAnalysis.</code><code class="sig-name descname">get_data</code><span class="sig-paren">(</span><em class="sig-param">data, drop_cols_exp=['subject', 'group', 'sample', 'index'], drop_cols_cli=['subject', 'group', 'biological_sample', 'index']</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/analytics_core/analytics/wgcnaAnalysis.html#get_data"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#analytics_core.analytics.wgcnaAnalysis.get_data" title="Permalink to this definition">¶</a></dt>
<dd><p>This function cleanes up and formats experimental and clinical data into similarly shaped dataframes.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#dict" title="(in Python v3.5)"><em>dict</em></a>) – dictionary with processed clinical and proteomics datasets.</p></li>
<li><p><strong>drop_cols_exp</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#list" title="(in Python v3.5)"><em>list</em></a>) – list of columns to drop from processed experimental (protemics/rna-seq/dna-seq) dataframe.</p></li>
<li><p><strong>drop_cols_cli</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#list" title="(in Python v3.5)"><em>list</em></a>) – list of columns to drop from processed clinical dataframe.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Dictionary with experimental and clinical dataframes (keys are the same as in the input dictionary).</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="analytics_core.analytics.wgcnaAnalysis.get_dendrogram">
<code class="sig-prename descclassname">analytics_core.analytics.wgcnaAnalysis.</code><code class="sig-name descname">get_dendrogram</code><span class="sig-paren">(</span><em class="sig-param">df</em>, <em class="sig-param">labels</em>, <em class="sig-param">distfun='euclidean'</em>, <em class="sig-param">linkagefun='ward'</em>, <em class="sig-param">div_clusters=False</em>, <em class="sig-param">fcluster_method='distance'</em>, <em class="sig-param">fcluster_cutoff=15</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/analytics_core/analytics/wgcnaAnalysis.html#get_dendrogram"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#analytics_core.analytics.wgcnaAnalysis.get_dendrogram" title="Permalink to this definition">¶</a></dt>
<dd><p>This function calculates the distance matrix and performs hierarchical cluster analysis on a set of dissimilarities and methods for analyzing it.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> – pandas dataframe with samples/subjects as index and features as columns.</p></li>
<li><p><strong>labels</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#list" title="(in Python v3.5)"><em>list</em></a>) – labels for the leaves of the tree.</p></li>
<li><p><strong>distfun</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – distance measure to be used (‘euclidean’, ‘maximum’, ‘manhattan’, ‘canberra’, ‘binary’, ‘minkowski’ or ‘jaccard’).</p></li>
<li><p><strong>linkagefun</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – hierarchical/agglomeration method to be used (‘single’, ‘complete’, ‘average’, ‘weighted’, ‘centroid’, ‘median’ or ‘ward’).</p></li>
<li><p><strong>div_clusters</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/functions.html#bool" title="(in Python v3.5)"><em>bool</em></a>) – dividing dendrogram leaves into clusters (True or False).</p></li>
<li><p><strong>fcluster_method</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – criterion to use in forming flat clusters.</p></li>
<li><p><strong>fcluster_cutoff</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/functions.html#int" title="(in Python v3.5)"><em>int</em></a>) – maximum cophenetic distance between observations in each cluster.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Dictionary of data structures computed to render the dendrogram. Keys: ‘icoords’, ‘dcoords’, ‘ivl’ and ‘leaves’. If div_clusters is used, it will also return a dictionary of each cluster and respective leaves.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="analytics_core.analytics.wgcnaAnalysis.get_clusters_elements">
<code class="sig-prename descclassname">analytics_core.analytics.wgcnaAnalysis.</code><code class="sig-name descname">get_clusters_elements</code><span class="sig-paren">(</span><em class="sig-param">linkage_matrix</em>, <em class="sig-param">fcluster_method</em>, <em class="sig-param">fcluster_cutoff</em>, <em class="sig-param">labels</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/analytics_core/analytics/wgcnaAnalysis.html#get_clusters_elements"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#analytics_core.analytics.wgcnaAnalysis.get_clusters_elements" title="Permalink to this definition">¶</a></dt>
<dd><p>This function implements the generation of flat clusters from an hierarchical clustering with the same interface as scipy.cluster.hierarchy.fcluster.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>linkage_matrix</strong> (<em>ndarray</em>) – hierarchical clustering encoded with a linkage matrix.</p></li>
<li><p><strong>fcluster_method</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – criterion to use in forming flat clusters (‘inconsistent’, ‘distance’, ‘maxclust’, ‘monocrit’, ‘maxclust_monocrit’).</p></li>
<li><p><strong>fcluster_cutoff</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/functions.html#float" title="(in Python v3.5)"><em>float</em></a>) – maximum cophenetic distance between observations in each cluster.</p></li>
<li><p><strong>labels</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#list" title="(in Python v3.5)"><em>list</em></a>) – labels for the leaves of the dendrogram.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A dictionary where keys are the cluster numbers and values are the dendrogram leaves.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="analytics_core.analytics.wgcnaAnalysis.filter_df_by_cluster">
<code class="sig-prename descclassname">analytics_core.analytics.wgcnaAnalysis.</code><code class="sig-name descname">filter_df_by_cluster</code><span class="sig-paren">(</span><em class="sig-param">df</em>, <em class="sig-param">clusters</em>, <em class="sig-param">number</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/analytics_core/analytics/wgcnaAnalysis.html#filter_df_by_cluster"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#analytics_core.analytics.wgcnaAnalysis.filter_df_by_cluster" title="Permalink to this definition">¶</a></dt>
<dd><p>Select only the members of a defined cluster.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> – pandas dataframe with samples/subjects as index and features as columns.</p></li>
<li><p><strong>clusters</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#dict" title="(in Python v3.5)"><em>dict</em></a>) – clusters dictionary from get_dendrogram function if div_clusters option was True.</p></li>
<li><p><strong>number</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/functions.html#int" title="(in Python v3.5)"><em>int</em></a>) – cluster number (key).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Pandas dataframe with all the features (columns) and samples/subjects belonging to the defined cluster (index).</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="analytics_core.analytics.wgcnaAnalysis.df_sort_by_dendrogram">
<code class="sig-prename descclassname">analytics_core.analytics.wgcnaAnalysis.</code><code class="sig-name descname">df_sort_by_dendrogram</code><span class="sig-paren">(</span><em class="sig-param">df</em>, <em class="sig-param">Z_dendrogram</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/analytics_core/analytics/wgcnaAnalysis.html#df_sort_by_dendrogram"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#analytics_core.analytics.wgcnaAnalysis.df_sort_by_dendrogram" title="Permalink to this definition">¶</a></dt>
<dd><p>Reorders pandas dataframe by index and according to the dendrogram list of leaf nodes labels.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> – pandas dataframe with the labels to be reordered as index.</p></li>
<li><p><strong>Z_dendrogram</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#dict" title="(in Python v3.5)"><em>dict</em></a>) – dictionary of data structures computed to render the dendrogram. Keys: ‘icoords’, ‘dcoords’, ‘ivl’ and ‘leaves’.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Reordered pandas dataframe.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="analytics_core.analytics.wgcnaAnalysis.get_percentiles_heatmap">
<code class="sig-prename descclassname">analytics_core.analytics.wgcnaAnalysis.</code><code class="sig-name descname">get_percentiles_heatmap</code><span class="sig-paren">(</span><em class="sig-param">df</em>, <em class="sig-param">Z_dendrogram</em>, <em class="sig-param">bydendro=True</em>, <em class="sig-param">bycols=False</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/analytics_core/analytics/wgcnaAnalysis.html#get_percentiles_heatmap"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#analytics_core.analytics.wgcnaAnalysis.get_percentiles_heatmap" title="Permalink to this definition">¶</a></dt>
<dd><p>This function transforms the absolute values in each row or column (option ‘bycols’) into relative values.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> – pandas dataframe with samples/subjects as index and features as columns.</p></li>
<li><p><strong>Z_dendrogram</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#dict" title="(in Python v3.5)"><em>dict</em></a>) – dictionary of data structures computed to render the dendrogram. Keys: ‘icoords’, ‘dcoords’, ‘ivl’ and ‘leaves’.</p></li>
<li><p><strong>bydendro</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/functions.html#bool" title="(in Python v3.5)"><em>bool</em></a>) – if labels should be ordered according to dendrogram list of leaf nodes labels set to True, otherwise set to False.</p></li>
<li><p><strong>bycols</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/functions.html#bool" title="(in Python v3.5)"><em>bool</em></a>) – relative values calculated across rows (samples) then set to False. Calculation performed across columns (features) set to True.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Pandas dataframe.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="analytics_core.analytics.wgcnaAnalysis.get_miss_values_df">
<code class="sig-prename descclassname">analytics_core.analytics.wgcnaAnalysis.</code><code class="sig-name descname">get_miss_values_df</code><span class="sig-paren">(</span><em class="sig-param">data</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/analytics_core/analytics/wgcnaAnalysis.html#get_miss_values_df"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#analytics_core.analytics.wgcnaAnalysis.get_miss_values_df" title="Permalink to this definition">¶</a></dt>
<dd><p>Proccesses pandas dataframe so missing values can be plotted in heatmap with specific color.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>data</strong> – pandas dataframe.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Pandas dataframe with missing values as integer 1, and originally valid values as NaN.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="analytics_core.analytics.wgcnaAnalysis.paste_matrices">
<code class="sig-prename descclassname">analytics_core.analytics.wgcnaAnalysis.</code><code class="sig-name descname">paste_matrices</code><span class="sig-paren">(</span><em class="sig-param">matrix1</em>, <em class="sig-param">matrix2</em>, <em class="sig-param">rows</em>, <em class="sig-param">cols</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/analytics_core/analytics/wgcnaAnalysis.html#paste_matrices"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#analytics_core.analytics.wgcnaAnalysis.paste_matrices" title="Permalink to this definition">¶</a></dt>
<dd><p>Takes two matrices with analog shapes and concatenates each value in matrix 1 with corresponding one in matrix 2, returning a single pandas dataframe.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>matrix1</strong> (<em>ndarray</em>) – input 1</p></li>
<li><p><strong>matrix2</strong> (<em>ndarray</em>) – input 2</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Pandas dataframe.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="analytics_core.analytics.wgcnaAnalysis.cutreeDynamic">
<code class="sig-prename descclassname">analytics_core.analytics.wgcnaAnalysis.</code><code class="sig-name descname">cutreeDynamic</code><span class="sig-paren">(</span><em class="sig-param">distmatrix</em>, <em class="sig-param">linkagefun='average'</em>, <em class="sig-param">minModuleSize=50</em>, <em class="sig-param">method='hybrid'</em>, <em class="sig-param">deepSplit=2</em>, <em class="sig-param">pamRespectsDendro=False</em>, <em class="sig-param">distfun=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/analytics_core/analytics/wgcnaAnalysis.html#cutreeDynamic"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#analytics_core.analytics.wgcnaAnalysis.cutreeDynamic" title="Permalink to this definition">¶</a></dt>
<dd><p>This function implements the R cutreeDynamic wrapper in Python, provinding an access point for methods of adaptive branh pruning of hierarchical clustering dendrograms.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> – pandas dataframe.</p></li>
<li><p><strong>distfun</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – distance measure to be used (‘euclidean’, ‘maximum’, ‘manhattan’, ‘canberra’, ‘binary’, ‘minkowski’ or ‘jaccard’).</p></li>
<li><p><strong>linkagefun</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – hierarchical/agglomeration method to be used (‘single’, ‘complete’, ‘average’, ‘weighted’, ‘centroid’, ‘median’ or ‘ward’).</p></li>
<li><p><strong>minModuleSize</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/functions.html#int" title="(in Python v3.5)"><em>int</em></a>) – minimum module size.</p></li>
<li><p><strong>method</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – method to use (‘hybrid’ or ‘tree’).</p></li>
<li><p><strong>deepSplit</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/functions.html#int" title="(in Python v3.5)"><em>int</em></a>) – provides a rough control over sensitivity to cluster splitting, the higher the value (with ‘hybrid’ method) or if True (with ‘tree’ method), the more and smaller modules.</p></li>
<li><p><strong>pamRespectsDendro</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/functions.html#bool" title="(in Python v3.5)"><em>bool</em></a>) – only used for method ‘hybrid’. Objects and small modules will only be assigned to modules that belong to the same branch in the dendrogram structure.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Numpy array of numerical labels giving assignment of objects to modules. Unassigned objects are labeled 0, the largest module has label 1, next largest 2 etc.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="analytics_core.analytics.wgcnaAnalysis.build_network">
<code class="sig-prename descclassname">analytics_core.analytics.wgcnaAnalysis.</code><code class="sig-name descname">build_network</code><span class="sig-paren">(</span><em class="sig-param">data</em>, <em class="sig-param">softPower=6</em>, <em class="sig-param">networkType='unsigned'</em>, <em class="sig-param">linkagefun='average'</em>, <em class="sig-param">method='hybrid'</em>, <em class="sig-param">minModuleSize=50</em>, <em class="sig-param">deepSplit=2</em>, <em class="sig-param">pamRespectsDendro=False</em>, <em class="sig-param">merge_modules=True</em>, <em class="sig-param">MEDissThres=0.4</em>, <em class="sig-param">verbose=0</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/analytics_core/analytics/wgcnaAnalysis.html#build_network"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#analytics_core.analytics.wgcnaAnalysis.build_network" title="Permalink to this definition">¶</a></dt>
<dd><p>Weighted gene network construction and module detection. Calculates co-expression similarity and adjacency, topological overlap matrix (TOM) and clusters features in modules.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> – pandas dataframe containing experimental data, with samples/subjects as rows and features as columns.</p></li>
<li><p><strong>softPower</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/functions.html#int" title="(in Python v3.5)"><em>int</em></a>) – soft-thresholding power.</p></li>
<li><p><strong>networkType</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – network type (‘unsigned’, ‘signed’, ‘signed hybrid’, ‘distance’).</p></li>
<li><p><strong>linkagefun</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – hierarchical/agglomeration method to be used (‘single’, ‘complete’, ‘average’, ‘weighted’, ‘centroid’, ‘median’ or ‘ward’).</p></li>
<li><p><strong>method</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – method to use (‘hybrid’ or ‘tree’).</p></li>
<li><p><strong>minModuleSize</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/functions.html#int" title="(in Python v3.5)"><em>int</em></a>) – minimum module size.</p></li>
<li><p><strong>pamRespectsDendro</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/functions.html#bool" title="(in Python v3.5)"><em>bool</em></a>) – only used for method ‘hybrid’. Objects and small modules will only be assigned to modules that belong to the same branch in the dendrogram structure.</p></li>
<li><p><strong>merge_modules</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/functions.html#bool" title="(in Python v3.5)"><em>bool</em></a>) – if True, very similar modules are merged.</p></li>
<li><p><strong>MEDissThres</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/functions.html#float" title="(in Python v3.5)"><em>float</em></a>) – maximum dissimilarity (i.e., 1-correlation) that qualifies modules for merging.</p></li>
<li><p><strong>verbose</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/functions.html#int" title="(in Python v3.5)"><em>int</em></a>) – integer level of verbosity. Zero means silent, higher values make the output progressively more and more verbose.</p></li>
</ul>
</dd>
<dt class="field-even">Paran int deepSplit</dt>
<dd class="field-even"><p>provides a rough control over sensitivity to cluster splitting, the higher the value (with ‘hybrid’ method) or if True (with ‘tree’ method), the more and smaller modules.</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>Tuple with TOM dissimilarity pandas dataframe, numpy array with module colors per experimental feature.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="analytics_core.analytics.wgcnaAnalysis.pick_softThreshold">
<code class="sig-prename descclassname">analytics_core.analytics.wgcnaAnalysis.</code><code class="sig-name descname">pick_softThreshold</code><span class="sig-paren">(</span><em class="sig-param">data</em>, <em class="sig-param">RsquaredCut=0.8</em>, <em class="sig-param">networkType='unsigned'</em>, <em class="sig-param">verbose=0</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/analytics_core/analytics/wgcnaAnalysis.html#pick_softThreshold"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#analytics_core.analytics.wgcnaAnalysis.pick_softThreshold" title="Permalink to this definition">¶</a></dt>
<dd><p>Analysis of scale free topology for multiple soft thresholding powers. Aids the user in choosing a proper soft-thresholding power for network construction.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> – pandas dataframe containing experimental data, with samples/subjects as rows and features as columns.</p></li>
<li><p><strong>RsquaredCut</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/functions.html#float" title="(in Python v3.5)"><em>float</em></a>) – desired minimum scale free topology fitting index R^2.</p></li>
<li><p><strong>networkType</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – network type (‘unsigned’, ‘signed’, ‘signed hybrid’, ‘distance’).</p></li>
<li><p><strong>verbose</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/functions.html#int" title="(in Python v3.5)"><em>int</em></a>) – integer level of verbosity. Zero means silent, higher values make the output progressively more and more verbose.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Estimated appropriate soft-thresholding power: the lowest power for which the scale free topology fit R^2 exceeds RsquaredCut.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.5/library/functions.html#int" title="(in Python v3.5)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="analytics_core.analytics.wgcnaAnalysis.identify_module_colors">
<code class="sig-prename descclassname">analytics_core.analytics.wgcnaAnalysis.</code><code class="sig-name descname">identify_module_colors</code><span class="sig-paren">(</span><em class="sig-param">matrix</em>, <em class="sig-param">linkagefun='average'</em>, <em class="sig-param">method='hybrid'</em>, <em class="sig-param">minModuleSize=30</em>, <em class="sig-param">deepSplit=2</em>, <em class="sig-param">pamRespectsDendro=False</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/analytics_core/analytics/wgcnaAnalysis.html#identify_module_colors"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#analytics_core.analytics.wgcnaAnalysis.identify_module_colors" title="Permalink to this definition">¶</a></dt>
<dd><p>Identifies co-expression modules and converts the numeric labels into colors.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>matrix</strong> – dissimilarity structure as produced by R.stats dist.</p></li>
<li><p><strong>minModuleSize</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/functions.html#int" title="(in Python v3.5)"><em>int</em></a>) – minimum module size.</p></li>
<li><p><strong>deepSplit</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/functions.html#int" title="(in Python v3.5)"><em>int</em></a>) – provides a rough control over sensitivity to cluster splitting, the higher the value (with ‘hybrid’ method) or if True (with ‘tree’ method), the more and smaller modules.</p></li>
<li><p><strong>pamRespectsDendro</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/functions.html#bool" title="(in Python v3.5)"><em>bool</em></a>) – only used for method ‘hybrid’. Objects and small modules will only be assigned to modules that belong to the same branch in the dendrogram structure.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Numpy array of strings with module color of each experimental feature.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="analytics_core.analytics.wgcnaAnalysis.calculate_module_eigengenes">
<code class="sig-prename descclassname">analytics_core.analytics.wgcnaAnalysis.</code><code class="sig-name descname">calculate_module_eigengenes</code><span class="sig-paren">(</span><em class="sig-param">data</em>, <em class="sig-param">modColors</em>, <em class="sig-param">softPower=6</em>, <em class="sig-param">dissimilarity=True</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/analytics_core/analytics/wgcnaAnalysis.html#calculate_module_eigengenes"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#analytics_core.analytics.wgcnaAnalysis.calculate_module_eigengenes" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates modules eigengenes to quantify co-expression similarity of entire modules.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> – pandas dataframe containing experimental data, with samples/subjects as rows and features as columns.</p></li>
<li><p><strong>modColors</strong> (<em>ndarray</em>) – array (numeric, character or a factor) attributing module colors to each feature in the experimental dataframe.</p></li>
<li><p><strong>softPower</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/functions.html#int" title="(in Python v3.5)"><em>int</em></a>) – soft-thresholding power.</p></li>
<li><p><strong>dissimilarity</strong> – calculates dissimilarity of module eigengenes.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Pandas dataframe with calculated module eigengenes. If dissimilarity is set to True, returns a tuple with two pandas dataframes, the first with the module eigengenes and the second with the eigengenes dissimilarity.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="analytics_core.analytics.wgcnaAnalysis.merge_similar_modules">
<code class="sig-prename descclassname">analytics_core.analytics.wgcnaAnalysis.</code><code class="sig-name descname">merge_similar_modules</code><span class="sig-paren">(</span><em class="sig-param">data</em>, <em class="sig-param">modColors</em>, <em class="sig-param">MEDissThres=0.4</em>, <em class="sig-param">verbose=0</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/analytics_core/analytics/wgcnaAnalysis.html#merge_similar_modules"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#analytics_core.analytics.wgcnaAnalysis.merge_similar_modules" title="Permalink to this definition">¶</a></dt>
<dd><p>Merges modules in co-expression network that are too close as measured by the correlation of their eigengenes.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> – pandas dataframe containing experimental data, with samples/subjects as rows and features as columns.</p></li>
<li><p><strong>modColors</strong> (<em>ndarray</em>) – array (numeric, character or a factor) attributing module colors to each feature in the experimental dataframe.</p></li>
<li><p><strong>verbose</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/functions.html#int" title="(in Python v3.5)"><em>int</em></a>) – integer level of verbosity. Zero means silent, higher values make the output progressively more and more verbose.</p></li>
</ul>
</dd>
<dt class="field-even">Para, float MEDissThres</dt>
<dd class="field-even"><p>maximum dissimilarity (i.e., 1-correlation) that qualifies modules for merging.</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>Tuple containing pandas dataframe with eigengenes of the new merged modules, and array with module colors of each expeirmental feature.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="analytics_core.analytics.wgcnaAnalysis.calculate_ModuleTrait_correlation">
<code class="sig-prename descclassname">analytics_core.analytics.wgcnaAnalysis.</code><code class="sig-name descname">calculate_ModuleTrait_correlation</code><span class="sig-paren">(</span><em class="sig-param">df_exp</em>, <em class="sig-param">df_traits</em>, <em class="sig-param">MEs</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/analytics_core/analytics/wgcnaAnalysis.html#calculate_ModuleTrait_correlation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#analytics_core.analytics.wgcnaAnalysis.calculate_ModuleTrait_correlation" title="Permalink to this definition">¶</a></dt>
<dd><p>Correlates eigengenes with external traits in order to identify the most significant module-trait associations.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df_exp</strong> – pandas dataframe containing experimental data, with samples/subjects as rows and features as columns.</p></li>
<li><p><strong>df_traits</strong> – pandas dataframe containing clinical data, with samples/subjects as rows and clinical traits as columns.</p></li>
<li><p><strong>MEs</strong> – pandas dataframe with module eigengenes.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Tuple with two pandas datafames, first the correlation between all module eigengenes and all clinical traits, second a dataframe with concatenated correlation and p-value used for heatmap annotation.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="analytics_core.analytics.wgcnaAnalysis.calculate_ModuleMembership">
<code class="sig-prename descclassname">analytics_core.analytics.wgcnaAnalysis.</code><code class="sig-name descname">calculate_ModuleMembership</code><span class="sig-paren">(</span><em class="sig-param">data</em>, <em class="sig-param">MEs</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/analytics_core/analytics/wgcnaAnalysis.html#calculate_ModuleMembership"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#analytics_core.analytics.wgcnaAnalysis.calculate_ModuleMembership" title="Permalink to this definition">¶</a></dt>
<dd><p>For each module, calculates the correlation of the module eigengene and the feature expression profile (quantitative measure of module membership (MM)).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> – pandas dataframe containing experimental data, with samples/subjects as rows and features as columns.</p></li>
<li><p><strong>MEs</strong> – pandas dataframe with module eigengenes.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Tuple with two pandas dataframes, one with module membership correlations and another with p-values.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="analytics_core.analytics.wgcnaAnalysis.calculate_FeatureTraitSignificance">
<code class="sig-prename descclassname">analytics_core.analytics.wgcnaAnalysis.</code><code class="sig-name descname">calculate_FeatureTraitSignificance</code><span class="sig-paren">(</span><em class="sig-param">df_exp</em>, <em class="sig-param">df_traits</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/analytics_core/analytics/wgcnaAnalysis.html#calculate_FeatureTraitSignificance"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#analytics_core.analytics.wgcnaAnalysis.calculate_FeatureTraitSignificance" title="Permalink to this definition">¶</a></dt>
<dd><p>Quantifies associations of individual experimental features with the measured clinical traits, by defining Feature Significance (FS) as the absolute value of the correlation between the feature and the trait.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df_exp</strong> – pandas dataframe containing experimental data, with samples/subjects as rows and features as columns.</p></li>
<li><p><strong>df_traits</strong> – pandas dataframe containing clinical data, with samples/subjects as rows and clinical traits as columns.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Tuple with two pandas dataframes, one with feature significance correlations and another with p-values.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="analytics_core.analytics.wgcnaAnalysis.get_FeaturesPerModule">
<code class="sig-prename descclassname">analytics_core.analytics.wgcnaAnalysis.</code><code class="sig-name descname">get_FeaturesPerModule</code><span class="sig-paren">(</span><em class="sig-param">data</em>, <em class="sig-param">modColors</em>, <em class="sig-param">mode='dictionary'</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/analytics_core/analytics/wgcnaAnalysis.html#get_FeaturesPerModule"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#analytics_core.analytics.wgcnaAnalysis.get_FeaturesPerModule" title="Permalink to this definition">¶</a></dt>
<dd><p>Groups all experimental features by the co-expression module they belong to.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> – pandas dataframe containing experimental data, with samples/subjects as rows and features as columns.</p></li>
<li><p><strong>modColors</strong> (<em>ndarray</em>) – array (numeric, character or a factor) attributing module colors to each feature in the experimental dataframe.</p></li>
<li><p><strong>mode</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#str" title="(in Python v3.5)"><em>str</em></a>) – type of the value returned by the function (‘dictionary’ or ‘dataframe’).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Depending on selected mode, returns a dictionary or dataframe with module color per experimental feature.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="analytics_core.analytics.wgcnaAnalysis.get_ModuleFeatures">
<code class="sig-prename descclassname">analytics_core.analytics.wgcnaAnalysis.</code><code class="sig-name descname">get_ModuleFeatures</code><span class="sig-paren">(</span><em class="sig-param">data</em>, <em class="sig-param">modColors</em>, <em class="sig-param">modules=[]</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/analytics_core/analytics/wgcnaAnalysis.html#get_ModuleFeatures"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#analytics_core.analytics.wgcnaAnalysis.get_ModuleFeatures" title="Permalink to this definition">¶</a></dt>
<dd><p>Groups and returns a list of the experimental features clustered in specific co-expression modules.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> – pandas dataframe containing experimental data, with samples/subjects as rows and features as columns.</p></li>
<li><p><strong>modColors</strong> (<em>ndarray</em>) – array (numeric, character or a factor) attributing module colors to each feature in the experimental dataframe.</p></li>
<li><p><strong>modules</strong> (<a class="reference external" href="https://docs.python.org/3.5/library/stdtypes.html#list" title="(in Python v3.5)"><em>list</em></a>) – list of module colors of interest.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>List of lists with experimental features in each selected module.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="analytics_core.analytics.wgcnaAnalysis.get_EigengenesTrait_correlation">
<code class="sig-prename descclassname">analytics_core.analytics.wgcnaAnalysis.</code><code class="sig-name descname">get_EigengenesTrait_correlation</code><span class="sig-paren">(</span><em class="sig-param">MEs</em>, <em class="sig-param">data</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/analytics_core/analytics/wgcnaAnalysis.html#get_EigengenesTrait_correlation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#analytics_core.analytics.wgcnaAnalysis.get_EigengenesTrait_correlation" title="Permalink to this definition">¶</a></dt>
<dd><p>Eigengenes are used as representative profiles of the co-expression modules, and correlation between them is used to quantify module similarity.
Clinical traits are added to the eigengenes to see how the traits fir into the eigengen network.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>MEs</strong> – pandas dataframe with module eigengenes.</p></li>
<li><p><strong>data</strong> – pandas dataframe containing clinical data, with samples/subjects as rows and clinical traits as columns.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Tuple with two pandas dataframes, one with features and traits recalculates module eigengenes dissimilarity, and another with all the overall correlations.</p>
</dd>
</dl>
</dd></dl>

</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="analytics_core.viz.html" class="btn btn-neutral float-right" title="Vizualization" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="analytics_core.html" class="btn btn-neutral float-left" title="Analytics Core" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2019, Alberto Santos, Ana Rita Colaço, Annelaura B. Nielsen

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>